#!/usr/bin/env Rscript

# clear the working environment and free up memory
rm(list=ls(all.names = TRUE))
gc()
repos='http://cran.us.r-project.org'

# check version, install and load libraries
v <- as.numeric(paste0(version$major, '.', 
                       strsplit(version$minor, "\\.")[[1]][1]))


if(v > 3.5 ){
  if(!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
    require(BiocManager)
  }
} else {
  print("Requires version 3.6 or above")
}


###########################################
######## load/install packages ############
###########################################

ipak <- function(pkg){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new.pkg)) 
    install.packages(new.pkg, dependencies = TRUE, repos = repos)
  sapply(pkg, require, character.only = TRUE)
}

pkgs <- c("ape", "data.table", "colorspace",  "devtools", "dplyr", "ggplot2", 
          "hrbrthemes", "lemon", "lubridate", "patchwork", "gghighlight",
          "scales", "tidytree", "tidyverse", "viridis", "countrycode",
          "plyr", "RColorBrewer", "cowplot", "grid", "gridExtra", "ggrepel",
          "ggthemes", "wesanderson", "ggthemes", "stringr", "ggpubr", "ggseqlogo",
          "phangorn", "showtext", "mutSignatures", "vcfR", "kableExtra")
ipak(pkg = pkgs)

# hrbrthemes::import_roboto_condensed()

ipak <- function(pkg){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new.pkg)) 
    BiocManager::install(new.pkg, dependencies = TRUE)
  sapply(pkg, require, character.only = TRUE)
}

pkgs <- c("Biostrings", "ComplexHeatmap", "ggtree", "ggmsa", "ggtreeExtra", 
          "ggnewscale", "tidytree", "treeio", "seqinr",
          "BSgenomeForge", "MutationalPatterns"
)
ipak(pkg = pkgs)


# additional fonts
# check the current search path for fonts
font_paths()
# List available font files in the search path
font_files()  
# Add desired font
font_add("Arial Narrow", "/System/Library/Fonts/Supplemental/Arial Narrow.ttf")
# list loaded fonts
font_families()


#
userDir <- path.expand("~")
baseDir <- file.path(userDir, "projects/pipelines/rvfvphylo")
projectDir <- file.path(userDir, "projects/vaccine-and-circulating-strains-analysis")

# output directories
outDir <- file.path(baseDir, "output-dir")
if (!dir.exists(outDir)) dir.create(outDir, recursive = TRUE)



setwd(outDir)

# global 
colors <- c("#9370DB", "#87CEFF", "#FF0000", "#551A8B", "#000033", 
            "#FF00B6", "#7AC5CD", "#FFD300", "#009FFF", "#8B1C62", 
            "#00FFBE", "#8B5A2B", "#1F9698", "#FFACFD", "#B1CC71")
df <- data.frame(lineage=toupper(letters[1:15]), color=colors)

prepare_metadata <- function(metadata) {
  
  # read metadata
  meta <- read.csv(metadata, sep="\t", header = TRUE)
  meta$taxa <- paste(meta$accession, meta$host,  meta$country, meta$date, sep="|")
  meta$date <- as.Date(as.character(meta$date), format = "%Y-%m-%d")
  meta$year <- sapply(strsplit(as.character(meta$date), "-", fixed = TRUE), `[`, 1)
  meta$sample_date <- decimal_date(meta$date)
  return(meta)
}




# function to plot nexus tree generated by treetime
plot_timetree <- function(tree, meta, root, cols, subtitle){
  
  #'@param tree tree object
  #'@param meta path to the file containing taxa metadata
  #'@param root decimal date value of the TMRCA of the tree (check with tempest)
  #'
  #'
  # root the tree
  # tree <- root(tree, "DQ380176|cow|KEN|1965-01-01",
  #              resolve.root = TRUE,
  #              edgelabel = TRUE)
  
  # make dataframe for clade nodes
  clades.df <- data.frame(clade=unique(meta$lineage), node=NA)
  
  # find the most recent common ancestor for each clade
  for (i in 1:length(clades.df$clade)) {
    clades.df$node[i] <- MRCA(
      tree,
      meta$taxa[meta$lineage == clades.df$clade[i]]
    )
  }
  
  # get mrsd
  mrsd <- meta[rev(order(as.Date(meta$date, format = "%Y-%m-%d"))),]$date[1]
  
  # plot preliminary tree
  gg.tree <- ggtree(tree, 
                    mrsd = mrsd, 
                    as.Date = TRUE, 
                    linetype=NA) %<+% meta +
    geom_highlight(data = clades.df,
                   aes(node=node, fill=clade),
                   alpha=1,
                   align="right",
                   extend=0.1,
                   show.legend=FALSE) +
    geom_tree(linewidth=0.5) +
    scale_x_date(limit=c(as.Date(decimal2Date(root)), 
                         as.Date(mrsd)), 
                 date_breaks= "14 year", 
                 date_labels = "%Y") +
    geom_tiplab(aes(label=accession), size=4)
  
  # order clades dataframe to match the tree
  clades.df <- clades.df[match(gg.tree$data %>% 
                                 dplyr::filter(isTip == "TRUE") %>% 
                                 arrange(y) %>% 
                                 pull(lineage) %>% 
                                 unique(),
                               clades.df$clade),]
  
  # add a column with alternating binary value
  clades.df$highlight <- rep(c(0,1), length.out=length(clades.df$clade))
  
  
  # highlight alternating colors for the clades
  p <- ggtree(tree, 
               mrsd = mrsd, 
               as.Date = FALSE, 
               size = 0.3) %<+% meta +
    geom_highlight(data = clades.df,
                   aes(node=node, fill=as.factor(highlight)),
                   alpha = 1,
                   align = "right",
                   extend = 0.02,
                   show.legend = FALSE
    ) +
    geom_tree(linewidth = 0.5) +
    geom_tippoint(aes(color = lineage), alpha = 1.0, size=24) +
    scale_color_manual(name = "Lineage",
                       values = cols,
                       breaks = sort(unique(meta$lineage)),
                       guide = guide_legend(override.aes = list(linetype = 0,
                                                                shape = 16,
                                                                color = cols))) +
    geom_tiplab(aes(subset=(strain_type %in% c('vaccine')),
                    label = paste0(strain)),
                align=FALSE,
                linesize=.5,
                linetype = "dashed",
                geom = "text",
                offset=0.01,
                size = 6,
                alpha=1,
                fontface=2
    ) +
    # geom_cladelab(data = clades.df,
    #               mapping = aes(node=node, label=clade, color=clade),
    #               fontsize = 16,
    #               barsize = 3.0,
    #               align = TRUE,
    #               offset = 0.008,
    #               offset.text = 0.005
    # ) +
    ggtitle("", subtitle = subtitle) +
    scale_fill_manual(values = c("#F5F5F5", "#ECECEC")) +
    theme_void(base_size = 32) +
    theme_tree2() +
    theme(text = element_text(family = "Helvetica"),
          plot.subtitle = element_text(size = 32, colour = "#000000", face = "bold"),
          plot.tag = element_text(colour = "#000000", face = "bold"),
          axis.text.x = element_text(size = 20, colour = "#000000", face = "bold"),
          axis.title.x = element_text(size = 22, colour = "#000000", face = "bold"),
          legend.text = element_text(size = 16, colour = "#000000", face = "bold"),
          legend.title = element_text(size = 18, colour = "#000000", face = "bold"),
          legend.position = c(.2, .6),
          # legend.position = "none",
          legend.key = element_blank(),
          legend.background = element_rect(color = "#000000",
                                           fill = "transparent",
                                           linewidth = 7,
                                           linetype = "blank")) +
    guides(fill = guide_legend(nrow = 15,
                               override.aes = list(size = 24),
                               title = "Lineage"))
  print(p)
  return(p)
}

plot_iqtree <- function(tree, metadata, cols, subtitle){
  #'
  #'
  #'
  #'
  
  meta <- metadata[match(tree@phylo$tip.label, metadata$taxa), ]
  print(all(tree@phylo$tip.label == meta$taxa))
  
  # make dataframe for clade nodes
  clades.df <- data.frame(clade=unique(meta$lineage), node=NA)
  
  # find the most recent common ancestor for each clade
  for (i in 1:length(clades.df$clade)) {
    clades.df$node[i] <- MRCA(
      tree,
      meta$taxa[meta$lineage == clades.df$clade[i]]
    )
  }
  
  # plot preliminary tree
  gg.tre <- ggtree(tree, 
                   as.Date = FALSE, 
                   linetype=NA) %<+% meta +
    geom_highlight(data = clades.df,
                   aes(node=node, fill=clade),
                   alpha=1,
                   align="right",
                   extend=0.1,
                   show.legend=FALSE)
  
  # order clades dataframe to match the tree
  clades.df <- clades.df[match(gg.tre$data %>% 
                                 dplyr::filter(isTip == "TRUE") %>% 
                                 arrange(y) %>% 
                                 pull(lineage) %>% 
                                 unique(),
                               clades.df$clade),]
  
  # add a column with alternating binary value
  clades.df$highlight <- rep(c(0,1), length.out=length(clades.df$clade))
  
  
  p <- 
    ggtree(tree, size = 0.3, color = "#000000") %<+% meta +
    geom_treescale(fontsize=4, linesize=1, offset=1) +
    geom_highlight(data = clades.df,
                   aes(node=node, fill=as.factor(highlight)),
                   alpha = 1,
                   align = "right",
                   extend = 0.002,
                   show.legend = FALSE
    ) +
    geom_tree(linewidth = 0.5) +
    geom_tippoint(aes(color = lineage), alpha = 1.0, size=24) +
    scale_color_manual(name = "Lineage",
                       values = cols,
                       breaks = sort(unique(meta$lineage)),
                       guide = guide_legend(override.aes = list(linetype = 0,
                                                                shape = 16,
                                                                color = cols))) +
    geom_tiplab(aes(subset=(strain_type %in% c('vaccine')),
                    label = paste0(strain)),
                align=FALSE,
                linesize=.5,
                geom = "text",
                linetype = "dashed",
                offset=0.001,
                size = 6,
                alpha=1,
                fontface=2
    ) +
    # geom_cladelab(data = clades.df,
    #               mapping = aes(node=node, label=clade, color=clade),
    #               fontsize = 15,
    #               barsize = 3,
    #               align = TRUE,
    #               offset = 0.004,
    #               offset.text = 0.01
    # ) +
    ggtitle("", subtitle = subtitle) +
    scale_fill_manual(values = c("#F5F5F5", "#ECECEC")) +
    geom_label2(aes(label=label, 
                    subset = !is.na(as.numeric(label)) & 
                      as.numeric(label) > 80, hjust=-.5, vjust=.5), size = 6) + 
    theme_void(base_size = 32) +
    theme_tree2() +
    theme(text = element_text(family = "Helvetica"),
          plot.margin = margin(1,1,1,1),
          plot.subtitle = element_text(size = 32, colour = "#000000", face = "bold"),
          plot.tag = element_text(colour = "#000000", face = "bold"),
          axis.text.x = element_text(size = 20, colour = "#000000", face = "bold"),
          axis.title.x = element_text(size = 22, colour = "#000000", face = "bold"),
          legend.text = element_text(size = 16, colour = "#000000", face = "bold"),
          legend.title = element_text(size = 18, colour = "#000000", face = "bold"),
          legend.position = c(.2, .6),
          # legend.position = "none",
          legend.key = element_blank(),
          legend.background = element_rect(color = "#000000",
                                           fill = "transparent",
                                           linewidth = 7,
                                           linetype = "blank")
    ) +
    guides(fill = guide_legend(nrow = 15,
                               override.aes = list(size = 24),
                               title = "Lineage"))
  print(p)
}



#################################### Large Segment  #############################
metadata <- file.path(baseDir, "output-dir/riftL/strain-types/riftL.traits.txt")
timetree <- file.path(baseDir, "output-dir/riftL/treetime/riftL_timetree.nexus")
treefile <- file.path(baseDir, "output-dir/riftL/iqtree/riftL.treefile")
metaL <- prepare_metadata(metadata = metadata)
metaL$lineage[metaL$accession == 'DQ375425'] <- 'D' # wrongly assigned as C
metaL$lineage[metaL$accession == 'DQ375433'] <- 'O' # wrongly assigned as L
metaL$lineage[metaL$accession == 'KX944854'] <- 'O' # wrongly assigned as L
metaL$lineage[metaL$accession == 'DQ375423'] <- 'G' # wrongly assigned as L
metaL$lineage[metaL$accession == 'DQ375422'] <- 'G' # wrongly assigned as L
metaL$lineage[metaL$accession == 'KJ782454'] <- 'G' # wrongly assigned as L

cols <- df[df$lineage %in% sort(unique(metaL$lineage)),]$color

# read nexus tree
tree.large <- read.nexus(timetree)
meta <- metaL[match(tree.large$tip.label, metaL$taxa), ]
all(tree.large$tip.label == meta$taxa)
plot.large.timetree <- plot_timetree(tree=tree.large, 
                            meta = meta, 
                            root = 1886.6015,
                            cols = cols,
                            subtitle = "RNA-dependent RNA polymerase (RdRp)")

plot.large.timetree
ggsave("ML-timetree-large.pdf", plot.large.timetree, 
       width = 15, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

# read IQTREE treefile
iqtreeLarge <- read.iqtree(treefile)
iqtreeLarge@phylo <- phangorn::midpoint(iqtreeLarge@phylo)
plot.large.iqtree <- plot_iqtree(tree = iqtreeLarge, 
                                 metadata = metaL,
                                 cols = cols,
                                 subtitle = "RNA-dependent RNA polymerase (RdRp)"
                                 )
plot.large.iqtree
ggsave("ML-iqtree-large.pdf", plot.large.iqtree, 
       width = 15, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

################################################################################

################################### Medium Segment #############################
metadata <- file.path(baseDir, "output-dir/riftM/strain-types/riftM.traits.txt")
timetree <- file.path(baseDir, "output-dir/riftM/treetime/riftM_timetree.nexus")
treefile <- file.path(baseDir, "output-dir/riftM/iqtree/riftM.treefile")
metaM <- prepare_metadata(metadata = metadata)
cols <- df[df$lineage %in% sort(unique(metaM$lineage)),]$color

# read nexus tree
tree.medium <- read.nexus(timetree)
meta <- metaM[match(tree.medium$tip.label, metaM$taxa), ]
all(tree.medium$tip.label == meta$taxa)
plot.medium.timetree <- plot_timetree(tree=tree.medium, 
                          meta = meta, 
                          root = 1902.9811,
                          cols = cols,
                          subtitle = "Glycoproteins (NSm/Gn/Gc)")

ggsave("ML-timetree-medium.pdf", plot.medium.timetree, 
       width = 15, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

# read IQTREE treefile
iqtreeMedium <- read.iqtree(treefile)
iqtreeMedium@phylo <- phangorn::midpoint(iqtreeMedium@phylo)
plot.medium.iqtree <- plot_iqtree(tree = iqtreeMedium, 
                                 metadata = metaM,
                                 cols = cols,
                                 subtitle = "Glycoproteins (NSm/Gn/Gc)"
)
ggsave("ML-iqtree-medium.pdf", plot.medium.iqtree, 
       width = 15, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

################################################################################


################################### Small Segment #############################
metadata <- file.path(baseDir, "output-dir/riftS-NSS/strain-types/riftS-NSS.traits.txt")
timetree <- file.path(baseDir, "output-dir/riftS-NSS/treetime/riftS-NSS_timetree.nexus")
treefile <- file.path(baseDir, "output-dir/riftS-NSS/iqtree/riftS-NSS.treefile")
metaS <- prepare_metadata(metadata = metadata)
cols <- df[df$lineage %in% sort(unique(metaS$lineage)),]$color

# read nexus tree
tree.nss <- read.nexus(timetree)
meta <- metaS[match(tree.nss$tip.label, metaS$taxa), ]
all(tree.nss$tip.label == meta$taxa)
plot.nss.timetree <- plot_timetree(tree=tree.nss, 
                            meta = meta, 
                            root = 1869.8172,
                            cols = cols,
                            subtitle = "Non-structural (NSs)"
                            )

ggsave("ML-timetree-nss.pdf", plot.nss.timetree, 
       width = 16, height = 26, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

# read IQTREE treefile
iqtreeNSS <- read.iqtree(treefile)
iqtreeNSS@phylo <- phangorn::midpoint(iqtreeNSS@phylo)
plot.nss.iqtree <- plot_iqtree(tree = iqtreeNSS, 
                                  metadata = metaS,
                                  cols = cols,
                                  subtitle = "Non-structural (NSs)"
)

ggsave("ML-iqtree-nss.pdf", plot.nss.iqtree, 
       width = 16, height = 26, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")


############################ nucleoprotein ######################
metadata <- file.path(baseDir, "output-dir/riftS-NP/strain-types/riftS-NP.traits.txt")
timetree <- file.path(baseDir, "output-dir/riftS-NP/treetime/riftS-NP_timetree.nexus")
treefile <- file.path(baseDir, "output-dir/riftS-NP/iqtree/riftS-NP.treefile")
metaS <- prepare_metadata(metadata = metadata)
cols <- df[df$lineage %in% sort(unique(metaS$lineage)),]$color

# read nexus tree
tree.np <- read.nexus(timetree)
meta <- metaS[match(tree.np$tip.label, metaS$taxa), ]
all(tree.np$tip.label == meta$taxa)
plot.np.timetree <- plot_timetree(tree=tree.nss, 
                                   meta = meta, 
                                   root = 1899.7875,
                                   cols = cols,
                                   subtitle = "Nucleoprotein (N)"
)

ggsave("ML-timetree-np.pdf", plot.np.timetree, 
       width = 15, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

# read IQTREE treefile
iqtreeNP <- read.iqtree(treefile)
iqtreeNP@phylo <- phangorn::midpoint(iqtreeNP@phylo)
plot.np.iqtree <- plot_iqtree(tree = iqtreeNP, 
                               metadata = metaS,
                               cols = cols,
                               subtitle = "Nucleoprotein (N)"
)


ggsave("ML-iqtree-np.pdf", plot.np.iqtree, 
       width = 16, height = 26, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

################################################################################



################################################################################


# save plots
plotOne <- (plot.large.timetree + theme(legend.position = "none") | plot.medium.timetree | plot.np.timetree + theme(legend.position = "none")) + 
  plot_layout(guides = 'collect', widths = c(1.1,1.2,1.5),
              heights = c(1,1,1)) + 
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 40, face = "bold"))
plotOne

ggsave("timetree-trees.pdf", plotOne, 
       width = 45, height = 27, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

ggsave("timetree-trees.png", plotOne, 
       width = 45, height = 27, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")

plotTwo <- (plot.large.iqtree + theme(legend.position = "none") | plot.medium.iqtree | plot.np.iqtree + theme(legend.position = "none")) + 
  plot_layout(guides = 'collect', 
              widths = c(1.1,1.2,1.2),
              heights = c(1,1,1)) + 
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 40,face = "bold"))

ggsave("iqtree-trees.pdf", plotTwo, 
       width = 36, height = 28, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

ggsave("iqtree-trees.png", plotTwo, 
       width = 36, height = 28, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")


################################################################################



################################################################################
############################## mutations plots #################################
################################################################################

prepare_tstv_data <- function(segment, alignment_dir){
  #'
  #'@param name description
  #'@param name description
  #'@param name description
  
  dir <- file.path(alignment_dir)
  fns <- file.path(dir, 
                   list.files(path = dir, 
                              pattern = "\\.parsed.bcftools.stats.csv$", 
                              recursive = TRUE))
  print(fns)
  
  data_list <- list()
  prefixes <- list()
  for (i in 1:length(fns)) {
    fn = fns[i]
    sample_name <- strsplit(basename(fn), ".", fixed=T)[[1]][3]
    print(strsplit(basename(fn), ".", fixed=T)[[1]][3])
    df <- read.csv(fn, sep=',', header=T)
    df$name <- sample_name
    data_list[[i]] <- df
  }
  d <- do.call(rbind, data_list)
  
  if (segment == 'Medium'){
    d$reference[d$name == 'DQ380193'] <- 'Smithburn'
    d$reference[d$name == 'DQ380208'] <- 'MP-12'
    d$reference[d$name == 'DQ380213'] <- 'Clone-13'
  }
  
  if (segment == 'Large'){
    d$reference[d$name == 'DQ375430'] <- 'Smithburn'
    d$reference[d$name == 'DQ375404'] <- 'MP-12'
    d$reference[d$name == 'DQ375417'] <- 'Clone-13'
    
  }
  
  if (segment == 'Small'){
    d$reference[d$name == 'DQ380157'] <- 'Smithburn'
    d$reference[d$name == 'DQ380154'] <- 'MP-12'
    d$reference[d$name == 'DQ380182'] <- 'Clone-13'
  }
  
  d <- d %>% dplyr::rename(
    'A->C'=A.C, 'A->G'=A.G, 'A->T'=A.T, 'C->A'=C.A, 'C->G'=C.G, 'C->T'=C.T,
    'G->A'=G.A, 'G->C'=G.C, 'G->T'=G.T, 'T->A'=T.A, 'T->C'=T.C, 'T->G'=T.G
  )
  
  
  mutations <- c('A->C', 'C->A', 'A->G', 'G->A', 'A->T', 'T->A',
                 'C->G', 'G->C', 'C->T', 'T->C', 'G->T', 'T->G')
  cols <- brewer.pal(12, "Paired")
  
  d <- d %>% dplyr::select(c(samples, reference, ts, tv, 
                             tstv, tsALT, tvALT, tstvALT, 
                             'A->C', 'C->A', 'A->G', 'G->A', 'A->T', 'T->A',
                             'C->G', 'G->C', 'C->T', 'T->C', 'G->T', 'T->G'))
  
  d.fq <- d
  print(names(d.fq)[9:20])
  d.fq$sum <- rowSums(d.fq[,names(d.fq)[9:20]])
  
  d.fq <- d.fq %>% 
    mutate(
      fAC = d$`A->C`/11979,
      fCA = d$`C->A`/11979,
      fAG = d$`A->G`/11979,
      fGA = d$`G->A`/11979,
      fAT = d$`A->T`/11979,
      fTA = d$`T->A`/11979,
      fCG = d$`C->G`/11979,
      fGC = d$`G->C`/11979,
      fCT = d$`C->T`/11979,
      fTC = d$`T->C`/11979,
      fGT = d$`G->T`/11979,
      fTG = d$`T->G`/11979
    )
  print(d.fq)
  
  d.melt <- reshape2::melt(d)
  dfq.melt <- reshape2::melt(d.fq)
  
  res <- list(d.melt, dfq.melt)
  return(res)
}


plot_substitutions <- function(data, reference, segment, gene){
  
  #'@description
  #'A short description...
  #'
  #'@param name description
  #'@param name description
  #'
  
  
  mutations <- c('A->C', 'C->A', 'A->G', 'G->A', 'A->T', 'T->A',
                 'C->G', 'G->C', 'C->T', 'T->C', 'G->T', 'T->G')
  cols <- brewer.pal(12, "Paired")
  
  data <- data[data$reference == reference,]
  
  
  ref_pos=(max(data$value)/4)+20
  seg_pos=(max(data$value)/4)-20
  gene_pos=(max(data$value)/4)-40
  
  
  sample = data[(data$variable == 'samples'),]$value
  ts = data[(data$variable == 'ts'),]$value
  tv = data[(data$variable == 'tv'),]$value
  tstv = data[(data$variable == 'tstv'),]$value
  
  ts.alt = data[(data$variable == 'tsALT'),]$value
  tv.alt = data[(data$variable == 'tvALT'),]$value
  tstv.alt = data[(data$variable == 'tstvALT'),]$value
  
  
  p <- ggplot(data = data[data$variable %in% mutations,], aes(x=variable, y=value, fill=variable)) +
    geom_bar(stat="identity", width = 0.7, position = 'dodge2') +
    ylab("SNV count") +
    ggtitle(paste("Vaccine", reference, "\n", "Segment", segment, "\n", "Gene", gene,  sep = " ")) +
    # theme_half_open(12) +
    theme_classic() +
    scale_fill_manual(name="Substitution",
                      values=c(cols, "black"),
                      breaks = mutations,
                      na.value = NA,
                      na.translate = F,
                      guide = guide_legend(override.aes = list(linetype = 0,
                                                               shape = 16,
                                                               color = cols))) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          text = element_text(family = "Helvetica"),
          plot.title = element_text(size = 15, hjust = 0.5, face = "bold", colour = "black"),
          plot.tag = element_text(size = 24, face = "bold", colour = "black"),
          axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          # axis.ticks.x = element_blank(),
          axis.text.y = element_text(size = 15, hjust = 1, vjust = 0.5, face = "bold", colour = "black"),
          axis.title.y = element_text(size = 15, face = "bold", colour = "black"),
          legend.text = element_text(size = 13, face = "bold", colour = "black"),
          legend.title = element_text(size = 14, face = "bold", colour = "black"),
          legend.key = element_blank()) +
    coord_capped_cart(bottom=capped_horizontal(c("both")), 
                      left=capped_vertical(c("both"))) +
    guides(fill = guide_legend(override.aes = list(size=12)))
  return(p)
}



plot_mutations_freqs <- function(data, reference, segment, gene){
  
  #'@description
  #'A short description...
  #'
  #'@param name description
  #'@param name description
  #'
  
  
  mutations <- c('fAC', 'fCA', 'fAG', 'fGA', 'fAT', 'fTA',
                 'fCG', 'fGC', 'fCT', 'fTC', 'fGT', 'fTG')
  cols <- brewer.pal(12, "Paired")
  
  data <- data[data$reference == reference,]
  print(data)
  
  
  ref_pos=(max(data$value)/4)+20
  seg_pos=(max(data$value)/4)-20
  gene_pos=(max(data$value)/4)-40
  
  
  sample = data[(data$variable == 'samples'),]$value
  ts = data[(data$variable == 'ts'),]$value
  tv = data[(data$variable == 'tv'),]$value
  tstv = data[(data$variable == 'tstv'),]$value
  
  ts.alt = data[(data$variable == 'tsALT'),]$value
  tv.alt = data[(data$variable == 'tvALT'),]$value
  tstv.alt = data[(data$variable == 'tstvALT'),]$value
  
  print(data[data$variable %in% mutations,])
  
  
  p <- ggplot(data = data[data$variable %in% mutations,], aes(x=variable, y=value, fill=variable)) +
    geom_bar(stat="identity", width = 0.7, position = 'dodge2') +
    ylab("Frequency") +
    ggtitle(paste("Vaccine", reference, "\n", "Segment", segment, "\n", "Gene", gene,  sep = " ")) +
    # theme_half_open(12) +
    theme_classic() +
    scale_fill_manual(name="Substitution",
                      values=c(cols, "black"),
                      breaks = mutations,
                      na.value = NA,
                      na.translate = F,
                      guide = guide_legend(override.aes = list(linetype = 0,
                                                               shape = 16,
                                                               color = cols))) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          text = element_text(family = "Helvetica"),
          plot.title = element_text(size = 15, hjust = 0.5, face = "bold", colour = "black"),
          plot.tag = element_text(size = 24, face = "bold", colour = "black"),
          axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          # axis.ticks.x = element_blank(),
          axis.text.y = element_text(size = 15, hjust = 1, vjust = 0.5, face = "bold", colour = "black"),
          axis.title.y = element_text(size = 15, face = "bold", colour = "black"),
          legend.text = element_text(size = 13, face = "bold", colour = "black"),
          legend.title = element_text(size = 14, face = "bold", colour = "black"),
          # axis.text.y = element_text(size = 15, hjust = 1, vjust = 0.5, face = "bold"),
          # axis.title.y = element_text(size = 15, face = "bold"),
          # legend.text = element_text(size = 13),
          # legend.title = element_text(size = 14),
          legend.key = element_blank()) +
    coord_capped_cart(bottom=capped_horizontal(c("both")), 
                      left=capped_vertical(c("both"))) +
    guides(fill = guide_legend(override.aes = list(size=12)))
  print(p)
  return(p)
}

############################### Large segment ##################################

inDirLarge <- file.path(baseDir, "output-dir/riftL/strain-types")
tstvLarge <- prepare_tstv_data(segment = "Large", 
                               alignment_dir = inDirLarge)

subsLargeMP12 <- plot_substitutions(data = tstvLarge[[1]], 
                         reference = "MP-12", 
                         segment = 'L', 
                         gene = "RdRp")
subsLargeMP12 <- subsLargeMP12 + theme(legend.position = "none")
subsLargeMP12

freqsLargeMP12 <- plot_mutations_freqs(data = tstvLarge[[2]], 
                                  reference = "MP-12", 
                                  segment = 'L', 
                                  gene = "RdRp")

freqsLargeMP12 <- freqsLargeMP12 + theme(legend.position = "none")
freqsLargeMP12

subsLargeClone13 <- plot_substitutions(data = tstvLarge[[1]], 
                         reference = "Clone-13", 
                         segment = 'L', 
                         gene = "RdRp")
subsLargeClone13

freqsLargeClone13 <- plot_mutations_freqs(data = tstvLarge[[2]], 
                                         reference = "Clone-13", 
                                         segment = 'L', 
                                         gene = "RdRp")
freqsLargeClone13

subsLargeSmithburn <- plot_substitutions(data = tstvLarge[[1]], 
                                         reference = "Smithburn", 
                                         segment = 'L', 
                                         gene = "RdRp")
subsLargeSmithburn <- subsLargeSmithburn + 
  theme(legend.position = "none")
subsLargeSmithburn

freqsLargeSmithburn <- plot_mutations_freqs(data = tstvLarge[[2]], 
                                           reference = "Smithburn", 
                                           segment = 'L', 
                                           gene = "RdRp")
freqsLargeSmithburn <- freqsLargeSmithburn + theme(legend.position = "none")


plots.large.tstv <- (subsLargeMP12 / subsLargeClone13 / subsLargeSmithburn) + 
  plot_layout(guides = 'collect') + 
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 32))

plots.large.tstv

ggsave("tstv-large.pdf", plots.large.tstv, 
       width = 12, height = 10, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")


############################## Medium segment ##################################

inDirMedium <- file.path(baseDir, "output-dir/riftM/strain-types")
tstvMedium <- prepare_tstv_data(segment = "Medium", 
                                alignment_dir = inDirMedium)
subsMediumMP12 <- plot_substitutions(data = tstvMedium[[1]], 
                                     reference = "MP-12", 
                                     segment = 'M', 
                                     gene = "NSm/Gn/Gc")
subsMediumMP12 <- subsMediumMP12 + theme(legend.position = "none")
subsMediumMP12

freqsMediumMP12 <- plot_mutations_freqs(data = tstvMedium[[2]], 
                                        reference = "MP-12", 
                                        segment = 'M', 
                                        gene = "NSm/Gn/Gc")
freqsMediumMP12 <- freqsMediumMP12 + theme(legend.position = "none")
freqsMediumMP12

subsMediumClone13 <- plot_substitutions(data = tstvMedium[[1]], 
                                        reference = "Clone-13", 
                                        segment = 'M', 
                                        gene = "NSm/Gn/Gc")
freqsMediumClone13 <- plot_mutations_freqs(data = tstvMedium[[2]], 
                                           reference = "Clone-13", 
                                           segment = 'M', 
                                           gene = "NSm/Gn/Gc")

subsMediumSmithburn <- plot_substitutions(data = tstvMedium[[1]], 
                                          reference = "Smithburn", 
                                          segment = 'M', 
                                          gene = "NSm/Gn/Gc")
subsMediumSmithburn <- subsMediumSmithburn + theme(legend.position = "none")

freqsMediumSmithburn <- plot_mutations_freqs(data = tstvMedium[[2]], 
                                             reference = "Smithburn", 
                                             segment = 'M', 
                                             gene = "NSm/Gn/Gc")
freqsMediumSmithburn <- freqsMediumSmithburn + theme(legend.position = "none")

plots.medium.tstv <- (subsMediumMP12 / subsMediumClone13 / subsMediumSmithburn) + 
  plot_layout(guides = 'collect') + 
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 32))

ggsave("tstv-medium.pdf", plots.medium.tstv, 
       width = 12, height = 10, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

############################ Small segment: NSs ################################

inDirNSS <- file.path(baseDir, "output-dir/riftS-NSS/strain-types")
tstvNSS <- prepare_tstv_data(segment = "Small", 
                             alignment_dir = inDirNSS)
subsNssMP12 <- plot_substitutions(data = tstvNSS[[1]], 
                                  reference = "MP-12", 
                                  segment = 'S', 
                                  gene = "NSs")
subsNssMP12 <- subsNssMP12 + theme(legend.position = "none")

freqsNssMP12 <- plot_mutations_freqs(data = tstvNSS[[2]], 
                                     reference = "MP-12", 
                                     segment = 'S', 
                                     gene = "NSs")
freqsNssMP12 <- freqsNssMP12 + theme(legend.position = "none")

subsNssClone13 <- plot_substitutions(data = tstvNSS[[1]], 
                                     reference = "Clone-13", 
                                     segment = 'S', 
                                     gene = "NSs")
freqsNssClone13 <- plot_mutations_freqs(data = tstvNSS[[2]], 
                                        reference = "Clone-13", 
                                        segment = 'S', 
                                        gene = "NSs")

subsNssSmithburn <- plot_substitutions(data = tstvNSS[[1]], 
                                       reference = "Smithburn", 
                                       segment = 'S', 
                                       gene = "NSs")
subsNssSmithburn <- subsNssSmithburn + theme(legend.position = "none")

freqsNssSmithburn <- plot_mutations_freqs(data = tstvNSS[[2]], 
                                          reference = "Smithburn", 
                                          segment = 'S', 
                                          gene = "NSs")
freqsNssSmithburn <- freqsNssSmithburn + theme(legend.position = "none")

plots.nss.tstv <- (subsNssMP12 / subsNssClone13 / subsNssSmithburn) + 
  plot_layout(guides = 'collect') + 
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 32))

ggsave("tstv-nss.pdf", plots.nss.tstv, 
       width = 12, height = 10, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

############################ Small segment: np ################################

inDirNP <- file.path(baseDir, "output-dir/riftS-NP/strain-types")
tstvNP <- prepare_tstv_data(segment = "Small", 
                            alignment_dir = inDirNP)
subsNpMP12 <- plot_substitutions(data = tstvNP[[1]], 
                                 reference = "MP-12", 
                                 segment = 'S', 
                                 gene = "NP")
subsNpMP12 <- subsNpMP12 + theme(legend.position = "none")

freqsNpMP12 <- plot_mutations_freqs(data = tstvNP[[2]], 
                                    reference = "MP-12", 
                                    segment = 'S', 
                                    gene = "NP")
freqsNpMP12 <- freqsNpMP12 + theme(legend.position = "none")

subsNpClone13 <- plot_substitutions(data = tstvNP[[1]], 
                                    reference = "Clone-13", 
                                    segment = 'S', 
                                    gene = "NP")
freqsNpClone13 <- plot_mutations_freqs(data = tstvNP[[2]], 
                                       reference = "Clone-13", 
                                       segment = 'S', 
                                       gene = "NP")

subsNpSmithburn <- plot_substitutions(data = tstvNP[[1]], 
                                      reference = "Smithburn", 
                                      segment = 'S', 
                                      gene = "NP")
subsNpSmithburn <- subsNpSmithburn + theme(legend.position = "none")

freqsNpSmithburn <- plot_mutations_freqs(data = tstvNP[[2]], 
                                         reference = "Smithburn", 
                                         segment = 'S', 
                                         gene = "NP")
freqsNpSmithburn <- freqsNpSmithburn + theme(legend.position = "none")

plots.np.tstv <- (subsNpMP12 / subsNpClone13 / subsNpSmithburn) + 
  plot_layout(guides = 'collect') + 
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 32))

ggsave("tstv-np.pdf", plots.np.tstv, 
       width = 12, height = 10, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

ggsave("tstv-np.png", plots.np.tstv, 
       width = 12, height = 10, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")



######################### per lineage per strain type comparison ###############

prepare_all_tstv_data <- function(segment, indir, gene){
  #'
  #'@param name description
  #'@param name description
  #'@param name description
  #'
  
  fns <- file.path(indir, list.files(path = indir, 
                                     pattern = "\\.all.csv$", 
                                     recursive = TRUE))
  data_list <- list()
  prefixes <- list()
  for (i in 1:length(fns)) {
    fn = fns[i]
    sample_name <- strsplit(basename(dirname(dirname(fn))), ".", fixed=T)[[1]][1]
    df <- read.csv(fn, sep=',', header=T)
    df$reference <- sample_name
    data_list[[i]] <- df
  }
  
  
  d <- bind_rows(data_list)
  d[c('Accession', 'Lineage', 'Strain')] <- str_split_fixed(d$SNP, ":", 3)
  
  if (gene == 'NP'){
    d <- d %>% dplyr::rename(
      'A->C'=AC, 'A->G'=AG, 'A->T'=AT, 'C->A'=CA, 'C->T'=CT,
      'G->A'=GA, 'G->C'=GC, 'G->T'=GT, 'T->A'=TA, 'T->C'=TC, 'T->G'=TG,
    )
  } else {
    d <- d %>% dplyr::rename(
      'A->C'=AC, 'A->G'=AG, 'A->T'=AT, 'C->A'=CA, 'C->G'=CG, 'C->T'=CT,
      'G->A'=GA, 'G->C'=GC, 'G->T'=GT, 'T->A'=TA, 'T->C'=TC, 'T->G'=TG,
    )
  }
  
  
  
  d <- d[order(d$Lineage),]
  if (segment == 'Medium'){
    d$reference[d$reference == 'DQ380208'] <- 'MP-12'
    d$reference[d$reference == 'DQ380213'] <- 'Clone-13'
    d$reference[d$reference == 'DQ380193'] <- 'Smithburn'
  }
  
  if (segment == 'Large'){
    d$reference[d$reference == 'DQ375430'] <- 'Smithburn'
    d$reference[d$reference == 'DQ375404'] <- 'MP-12'
    d$reference[d$reference == 'DQ375417'] <- 'Clone-13'
  }
  
  if (segment == 'Small'){
    d$reference[d$reference == 'DQ380157'] <- 'Smithburn'
    d$reference[d$reference == 'DQ380154'] <- 'MP-12'
    d$reference[d$reference == 'DQ380182'] <- 'Clone-13'
    
  }
  
  # boxplots
  print(compare_means(Ti.Tv ~ Lineage,  data = d,
                      ref.group = ".all.", method = "wilcox"))
  return(d)
  
}


plot_tstv_per_lineage <- function(data, reference, cols, segment, gene){
  
  data <- data[(data$reference == reference),]
  p <- ggboxplot(data, x = "Lineage", y = "Ti.Tv", color = "Lineage") +
    scale_color_manual(name="Lineage",
                       values=cols,
                       breaks = sort(unique(data$Lineage)),
                       na.value = NA,
                       na.translate = F) +
    ylab("Mean Ti/Tv Ratio") +
    ggtitle(paste("Vaccine", reference, "\n", "Segment", segment, "\n", "Gene", gene,  sep = " ")) +
    # ggtitle(paste(segment, "-", gene, "Ref:", reference, sep = " ")) +
    # ggtitle(paste("Vaccine:", reference,";", "Segment:", segment, ";", "Gene:", gene,  sep = "")) +
    theme_classic() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          text = element_text(family = "Helvetica"),
          plot.title = element_text(size = 15, hjust = 0.5, face = "bold", colour = "black"),
          plot.tag = element_text(size = 24, face = "bold", colour = "black"),
          axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          # axis.ticks.x = element_blank(),
          axis.text.y = element_text(size = 15, hjust = 1, vjust = 0.5, face = "bold", "black"),
          axis.title.y = element_text(size = 16, face = "bold", colour = "black"),
          legend.text = element_text(size = 13, face = "bold", colour = "black"),
          legend.title = element_text(size = 14, face = "bold", colour = "black"),
          legend.key = element_blank()) +
    coord_capped_cart(bottom=capped_horizontal(c("both")), 
                      left=capped_vertical(c("both"))) +
    # theme(panel.grid.major = element_blank(),
    #       panel.grid.minor = element_blank(),
    #       text = element_text(family = "Helvetica"),
    #       plot.title = element_text(size = 16, hjust = 0.5),
    #       plot.tag = element_text(size = 24),
    #       axis.title.x = element_blank(),
    #       axis.text.x = element_blank(),
    #       axis.ticks.x = element_blank(),
    #       axis.text.y = element_text(size = 15, hjust = 1, vjust = 0.5),
    #       axis.title.y = element_text(size = 16),
    #       legend.text = element_text(size = 13),
  #       legend.title = element_text(size = 14),
  #       legend.key = element_blank()) +
  # theme(panel.grid.major = element_blank(),
  #       panel.grid.minor = element_blank(),
  #       axis.title.x = element_blank(),
  #       axis.text.x = element_blank(),
  #       axis.ticks.x = element_blank(),
  #       axis.text.y = element_text(size = 18, hjust = 1, vjust = 0.5),
  #       axis.title.y = element_text(size = 20),
  #       legend.text = element_text(size = 15),
  #       legend.title = element_text(size = 16),
  #       legend.key = element_blank(),
  #       # plot.title = element_text(hjust = 0.5, vjust = -25.0, size = 15)
  # ) +
  stat_compare_means(label = "p.signif", method = "wilcox", ref.group = ".all.")
  print(p)
}


subsLargeMP12 <- subsLargeMP12 + 
  theme(legend.position = "none")
subsLargeSmithburn <- subsLargeSmithburn + 
  theme()
subsLargeClone13 <- subsLargeClone13 + 
  theme(legend.position = "left")

subsMediumMP12 <- subsMediumMP12 +
  theme(legend.position = "none",
        axis.title.y = element_blank())
subsMediumSmithburn <- subsMediumSmithburn +
  theme(legend.position = "none",
        axis.title.y = element_blank())
subsMediumClone13 <- subsMediumClone13 +
  theme(legend.position = "none",
        axis.title.y = element_blank())

subsNssMP12 <- subsNssMP12 + 
  theme(legend.position = "none",
        axis.title.y = element_blank())
subsNssSmithburn <- subsNssSmithburn + 
  theme(legend.position = "none",
        axis.title.y = element_blank())
subsNssClone13 <- subsNssClone13 + 
  theme(legend.position = "none",
        axis.title.y = element_blank())

subsNpMP12 <- subsNpMP12 + 
  theme(legend.position = "none",
        axis.title.y = element_blank())
subsNpSmithburn <- subsNpSmithburn + 
  theme(legend.position = "none",
        axis.title.y = element_blank())
subsNpClone13 <- subsNpClone13 + 
  theme(legend.position = "none",
        axis.title.y = element_blank())




combined.all.tstv.plots <- 
  ((subsLargeMP12 / subsLargeSmithburn / subsLargeClone13) | 
     (subsMediumMP12 / subsMediumSmithburn / subsMediumClone13) | 
     (subsNssMP12 / subsNssSmithburn / subsNssClone13) | 
     (subsNpMP12 / subsNpSmithburn / subsNpClone13)) +
  plot_layout(guides = 'collect') +
  plot_annotation(tag_levels = "A")

ggsave("Figure3-combined-all.tstv.pdf", 
       combined.all.tstv.plots,
       width = 18, height = 10, 
       units = "in", limitsize = FALSE,
       dpi = 300, bg="white", device = "pdf")

ggsave("Figure3-combined-all.tstv.png", 
       combined.all.tstv.plots,
       width = 18, height = 10, 
       units = "in", limitsize = FALSE,
       dpi = 600, bg="white", device = "png")

####################### mutation maps #########################


codeMutationTags<- function(count, thre){
  over<- names(count[which(as.numeric(count)>=thre)])
  print(over)
  over<- strsplit(over, "(?=[A-Za-z])(?<=[0-9])|(?=[0-9])(?<=[A-Za-z])", perl=TRUE)
  over<- matrix(unlist(over),ncol=3,byrow=T)
  positions<- over[,2]
  over<- cbind(over[,2], ": ", over[,1], " to ", over[,3])
  over<- apply(over, MARG=1, FUN=function(X) { paste(X, collapse='') })
  over<- data.frame(x_pos=as.numeric(positions), y_pos=thre/2, tag=as.character(over), stringsAsFactors=FALSE)
  if (dim(over)[1] >= 1){
    return(over)
  }
}

# large segment
# 1 antelope
# 2 bat
# 3 buffalo
# 67 cow
# 170 human
# 22 mosquito
# 16 sheep

# medium segment
# 1 antelope
# 2 bat
# 3 buffalo
# 73 cow
# 168 human
# 21 mosquito
# 15 sheep

# nss
# 1 antelope
# 3 buffalo
# 76 cow
# 196 human
# 24 mosquito
# 15 sheep

plot_mutation_map <- function(fpath, threshold, lineage, aa_size, step, palette, subtitle){
  
  mutation_df <- read.csv(fpath, header = TRUE)
  df.mut.s <- mutation_df %>% 
    dplyr::group_by(host,snps) %>% 
    dplyr::summarise(count=n()) %>%
    dplyr::arrange(desc(count))
  print(df.mut.s)

  lineage_df <- mutation_df %>% 
    dplyr::select(snps, positions, host) %>% 
    dplyr::rename("mutation" = snps, "location" = positions)
  # lineage_df <- lineage_df[lineage_df$host == host_organism,]
  
  
  tags <- codeMutationTags(table(lineage_df$mutation), thre=threshold)

  if (dim(tags)[1] >= 1) {
    tags <- tags[order(tags$x_pos,decreasing=FALSE),]
    table <- table(lineage_df$location)
    table <- data.frame(count=as.numeric(table),location=as.numeric(names(table)))
    to_label <- table[table$location %in% tags$x_pos,]
    to_label <- merge(to_label, tags, by.x = "location", by.y = "x_pos")
    to_label$tag <- paste0(sapply(strsplit(gsub(":", "", to_label$tag), " "), `[`, 2),
                           sapply(strsplit(gsub(":", "", to_label$tag), " "), `[`, 1),
                           sapply(strsplit(gsub(":", "", to_label$tag), " "), `[`, 4))
    
    rows <- dim(to_label)[1]
    p <- ggplot(table, aes(x = location, y=count)) +
      geom_bar(data=to_label, aes(x=location, y=count, fill=tag), 
               stat="identity", width = 2.5, alpha=1) +
      xlab("Genome site (Amino acid)")+ ylab("No. of genomes") +
      ggtitle("", subtitle = subtitle) +
      scale_x_continuous(limits = c(0, aa_size),breaks = seq(0, aa_size, by = step)) +
      geom_hline(yintercept=threshold, linetype="dashed", color = "#545454") +
      geom_point(aes(x = location, y = count, colour=count), 
                 data=to_label) +
      scale_colour_continuous_sequential(name="No. of Sequences",
                                         palette = palette,
                                         rev = TRUE,
                                         alpha=0.5,
                                         guide = "colourbar") +
    geom_label_repel(data = to_label,
                     aes(label=tag, fill=tag),
                     size = 6,
                     colour = "#000000",
                     fontface = "bold",
                     nudge_x = 0.25,
                     nudge_y = 0.25,
                     show.legend = FALSE,
                     max.overlaps = 20,
                     angle = 0) +
      scale_fill_discrete_sequential(palette = palette, 
                                     rev = TRUE, alpha=0.5) +
      theme_classic() +
      theme_classic(base_size = 30) +
      theme(
        text=element_text(family="Helvetica"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "#000000", linewidth = 0.6),
        axis.ticks.y = element_line(colour = "#000000", linewidth = 0.6),
        axis.ticks.x = element_blank(),
        plot.subtitle = element_text(colour = "#000000", face = "bold"),
        plot.tag = element_text(colour = "#000000", face = "bold"),
        axis.text = element_text(colour = "#000000", face = "bold"),
        axis.title = element_text(colour = "#000000", face = "bold"),
        legend.position = "none",
        # legend.text = element_text(size = 18),
        # legend.title = element_text(size = 20),
        legend.key = element_blank()) +
      guides(colour = guide_legend(nrow = rows, 
                                   override.aes = list(size = 12)))
    print(p)
  }
  l <- list(lineage_df, p)
  return(l)
}

############################### Large segment ##################################

largeSingletonMP12File <- file.path(baseDir, "output-dir/riftL/strain-types/riftL.strain_type.DQ375404.mutations.per.strain.singleton.csv")
largeSingletonClone13File <- file.path(baseDir, "output-dir/riftL/strain-types/riftL.strain_type.DQ375417.mutations.per.strain.singleton.csv")
largeSingletonSmithburnFile <- file.path(baseDir, "output-dir/riftL/strain-types/riftL.strain_type.DQ375430.mutations.per.strain.singleton.csv")

# threshold <- round(90/100*282)
largeMutMapMP12 <- plot_mutation_map(fpath = largeSingletonMP12File,
                              threshold = round(90/100*281),
                              lineage = "MP-12",
                              aa_size = 2092, 
                              step = 200,
                              palette = "ag_GrnYl",
                              subtitle = "RdRp, vaccine: MP-12")

plotLargeMutMapMP12 <- largeMutMapMP12[[2]] + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotLargeMutMapMP12

largeMutMapClone13 <- plot_mutation_map(fpath = largeSingletonClone13File,
                              threshold = round(80/100*281),
                              lineage = "Clone 13",
                              aa_size = 2092, 
                              step = 200,
                              palette = "YlGnBu",
                              subtitle = "RdRp, vaccine: Clone 13")

plotLargeMutMapClone13 <- largeMutMapClone13[[2]] + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotLargeMutMapClone13

largeMutMapSmithburn <- plot_mutation_map(fpath = largeSingletonSmithburnFile,
                              threshold = round(90/100*281),
                              lineage = "Smithburn",
                              aa_size = 2092, 
                              step = 200,
                              palette = "Batlow",
                              subtitle = "RdRp, vaccine: Smithburn")
plotLargeMutMapSmithburn <- largeMutMapSmithburn[[2]] + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotLargeMutMapSmithburn



############################### Medium segment #################################

mediumSingletonMP12File <- file.path(baseDir, "output-dir/riftM/strain-types/riftM.strain_type.DQ380208.mutations.per.strain.singleton.csv")
mediumSingletonClone13File <- file.path(baseDir, "output-dir/riftM/strain-types/riftM.strain_type.DQ380213.mutations.per.strain.singleton.csv")
mediumSingletonSmithburnFile <- file.path(baseDir, "output-dir/riftM/strain-types/riftM.strain_type.DQ380193.mutations.per.strain.singleton.csv")


mediumMutMapMP12 <- plot_mutation_map(fpath = mediumSingletonMP12File,
                              threshold = round(90/100*284),
                              lineage = "MP-12",
                              aa_size = 1197, 
                              step = 200,
                              palette = "YlGnBu",
                              subtitle = "NSm/Gn/Gc, vaccine: MP-12")
plotMediumMutMapMP12 <- mediumMutMapMP12[[2]] + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank())
plotMediumMutMapMP12

mediumMutMapClone13 <- plot_mutation_map(fpath = mediumSingletonClone13File,
                              threshold = round(90/100*284),
                              lineage = "Clone 13",
                              aa_size = 1197, 
                              step = 200,
                              palette = "ag_GrnYl",
                              subtitle = "NSm/Gn/Gc, vaccine: Clone-13")
plotMediumMutMapClone13 <- mediumMutMapClone13[[2]] + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotMediumMutMapClone13

plotMediumMutMapSmithburn <- plot_mutation_map(fpath = mediumSingletonSmithburnFile,
                              threshold = round(90/100*284),
                              lineage = "Smithburn",
                              aa_size = 1197, 
                              step = 200,
                              palette = "Batlow",
                              subtitle = "NSm/Gn/Gc, vaccine: Smithburn")
plotMediumMutMapSmithburn <- plotMediumMutMapSmithburn[[2]] + 
  theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotMediumMutMapSmithburn

############################# Small segment: NSs ###############################

nssSingletonMP12File <- file.path(baseDir, "output-dir/riftS-NSS/strain-types/riftS-NSS.strain_type.DQ380154.mutations.per.strain.singleton.csv")
nssSingletonClone13File <- file.path(baseDir, "output-dir/riftS-NSS/strain-types/riftS-NSS.strain_type.DQ380182.mutations.per.strain.singleton.csv")
nssSingletonSmithburnFile <- file.path(baseDir, "output-dir/riftS-NSS/strain-types/riftS-NSS.strain_type.DQ380157.mutations.per.strain.singleton.csv")


nssMutMapMP12 <- plot_mutation_map(fpath = nssSingletonMP12File,
                                threshold = round(90/100*316),
                                lineage = "MP-12",
                                aa_size = 265, 
                                step = 50,
                                palette = "YlGnBu",
                                subtitle = "NSs, vaccine: MP-12")

plotNssMutMapMP12 <- nssMutMapMP12[[2]] + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank())
plotNssMutMapMP12

nssMutMapClone13 <- plot_mutation_map(fpath = nssSingletonClone13File,
                                threshold = round(90/100*316),
                                lineage = "Clone 13",
                                aa_size = 265, 
                                step = 50,
                                palette = "ag_GrnYl",
                                subtitle = "NSs, vaccine: Clone-13")
plotNssMutMapClone13 <- nssMutMapClone13[[2]] + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotNssMutMapClone13

nssMutMapSmithburn <- plot_mutation_map(fpath = nssSingletonSmithburnFile,
                                threshold = round(90/100*316),
                                lineage = "Smithburn",
                                aa_size = 265, 
                                step = 50,
                                palette = "Batlow",
                                subtitle = "NSs, vaccine: Smithburn")
plotNssMutMapSmithburn <- nssMutMapSmithburn[[2]] + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotNssMutMapSmithburn

############################## Small segment: NP ###############################

npSingletonMP12File <- file.path(baseDir, "output-dir/riftS-NP/strain-types/riftS-NP.strain_type.DQ380154.mutations.per.strain.singleton.csv")
npSingletonClone13File <- file.path(baseDir, "output-dir/riftS-NP/strain-types/riftS-NP.strain_type.DQ380182.mutations.per.strain.singleton.csv")
npSingletonSmithburnFile <- file.path(baseDir, "output-dir/riftS-NP/strain-types/riftS-NP.strain_type.DQ380157.mutations.per.strain.singleton.csv")


npMutMapMP12 <- plot_mutation_map(fpath = npSingletonMP12File,
                                  threshold = round(90/100*316),
                                  lineage = "MP-12",
                                  aa_size = 265, 
                                  step = 50,
                                  palette = "YlGnBu",
                                  subtitle = "NP, vaccine: MP-12")

plotNpMutMapMP12 <- npMutMapMP12[[2]] + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank())
plotNpMutMapMP12

npMutMapClone13 <- plot_mutation_map(fpath = npSingletonClone13File,
                                     threshold = round(5/100*316),
                                     lineage = "Clone 13",
                                     aa_size = 265, 
                                     step = 50,
                                     palette = "ag_GrnYl",
                                     subtitle = "NP, vaccine: Clone-13")
plotNpMutMapClone13 <- npMutMapClone13[[2]] + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotNpMutMapClone13

npMutMapSmithburn <- plot_mutation_map(fpath = npSingletonSmithburnFile,
                                       threshold = round(90/100*316),
                                       lineage = "Smithburn",
                                       aa_size = 265, 
                                       step = 50,
                                       palette = "Batlow",
                                       subtitle = "NP, vaccine: Smithburn")
plotNpMutMapSmithburn <- npMutMapSmithburn[[2]] + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank(),
  axis.title.x = element_blank(),
  axis.text.x = element_blank()
)
plotNpMutMapSmithburn

mutation.plots <- (
  (plotLargeMutMapMP12 | plotLargeMutMapClone13 | plotLargeMutMapSmithburn) / 
    (plotMediumMutMapMP12 | plotMediumMutMapClone13 | plotMediumMutMapSmithburn) / 
    (plotNssMutMapMP12 | plotNssMutMapClone13 | plotNssMutMapSmithburn) / 
    (plotNpMutMapMP12 | plotNpMutMapClone13 | plotNpMutMapSmithburn)
  ) +
  plot_layout(guides = 'collect', widths = c(1,8,12), heights = c(2,2,2,2)) +
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 32, face = "bold", colour = "black"))
#mutation.plots

mut.plots <- (
  (plotLargeMutMapMP12 | plotLargeMutMapSmithburn) / 
    (plotMediumMutMapMP12 | plotMediumMutMapSmithburn) / 
    (plotNssMutMapMP12 | plotNssMutMapSmithburn) / 
    (plotNpMutMapMP12 | plotNpMutMapSmithburn)) +
  plot_layout(guides = 'collect', widths = c(1,8,12), 
              heights = c(2,2,2,2)) +
  plot_annotation(tag_levels = "A") & 
  theme(plot.tag = element_text(size = 32, face = "bold", colour = "black"))

# mut.plots


ggsave("mutation-plots.pdf", mut.plots, 
       width = 28, height = 21, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)

ggsave("mutation-plots.png", mut.plots, 
       width = 28, height = 21, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")

#### plot proportions

plot_mutations <- function(file, snps_count, reference, gene, palette){
  #'
  #'
  #'
  #'
  #'
  
  df <- read.csv(file)
  df$host <- str_to_title(df$host)
  df1 <- df %>% 
    dplyr::group_by(host, snps, positions) %>% 
    dplyr::summarise(count=n()) 
  
  df2 <- df %>% dplyr::group_by(host, snps, positions) %>% 
          dplyr::summarise(count=n()) %>% 
          dplyr::arrange(desc(count))
  
  data_list <- list()
  df3 <- NULL
  for (i in 1:length(unique(df2$host))){
    df3$host <- unique(df2$host)[i]
    df3$total <- df2[df2$host == unique(df2$host)[i],]$count[[1]]
    data_list[[i]] <- df3
  }
  dat <- bind_rows(data_list)
  # print(dat)
  
  
  # print(df2 %>% dplyr::group_by(host, count) %>% 
  #         dplyr::summarise_at(vars(c(2)),
  #                             list(
  #                               Max=max)))

  # print(df %>% 
  #         dplyr::group_by(snps, positions) %>% 
  #         dplyr::summarise(count=n()) %>% 
  #         dplyr::arrange(desc(count)) %>% 
  #         summarise_at(vars(c(2)), 
  #                      list(
  #                        Min = min, 
  #                        Mean = mean,
  #                        Median = median,
  #                        Max = max, 
  #                        Sd = sd,
  #                        cv = (~sd(.)/mean(.) * 100))
  #         ) %>%
  #         dplyr::mutate_if(is.numeric, round, digits=2) %>%
  #         dplyr::arrange(desc(Mean)))
  # get abundant snps
  data <- df1[df1$count > snps_count, ]
  data <- data[order(data$positions),]
  data <- dplyr::left_join(data, dat, by="host")
  data$percentage <- data$count/data$total*100
  data <- data |> dplyr::mutate_if(is.numeric, round, digits=1) 
  data <- data |> dplyr::mutate(perc = paste0(sprintf("%4.1f", count / total * 100), "%"))
  # print(data |> 
  #         dplyr::group_by(positions) |> 
  #         dplyr::summarise(count=n()) |> 
  #         dplyr::filter(count > 4), 
  #       n=100)
  # 
  
  print(data %>%
          dplyr::group_by(snps, positions) %>%
          dplyr::summarise(count=n()) %>%
          dplyr::filter(count >= 3) %>% 
          dplyr::arrange(desc(positions)), 
        n=100)
  
  p <- ggplot(data, aes(x = count, 
                        y = reorder(snps, sort(as.numeric(positions))), 
                        fill = snps)) +
    geom_col() +
    geom_label(data=data[data$percentage >=50,],
      aes(label = perc), 
      hjust = 1, 
      nudge_x = -.5,
      size = 7, 
      fontface = "bold", 
      family = "Arial Narrow",
      ## turn into white box without outline
      fill = "white", label.size = 0
    ) +
    ylab("Mutations") +
    xlab("No. of sequences") +
    ggtitle(paste(gene, ", vaccine: ", reference, sep = "")) +
    facet_grid(facets=. ~ host, 
               labeller=labeller(drv=labels), 
               scales = "free" ) +
    scale_fill_discrete_diverging(palette = palette, rev = TRUE, alpha=1.0) +
    scale_x_continuous(expand = c(.01, .01)) +
    theme_classic(base_size = 30) +
    theme(
      text=element_text(family="Arial Narrow"),
      axis.line = element_line(colour = "#000000", linewidth = 0.4),
      axis.ticks = element_line(colour = "#000000", linewidth = 0.4),
      axis.title = element_text(colour = "#000000", face = "bold"),
      panel.grid.major = element_line(colour = "#B3B3B3", linewidth = 0.2),
      panel.grid.minor = element_blank(),
      panel.grid.major.y = element_blank(),
      panel.grid.minor.y = element_blank(),
      panel.spacing = unit(0.75, "cm", data = NULL),
      strip.background = element_blank(),
      strip.text.x = element_text(colour = "#000000", face = "bold", hjust = 0),
      plot.title = element_text(colour = "#000000", face = "bold"),
      axis.text = element_text(colour = "#000000", face = "bold"),
      legend.position = "none",
      plot.margin = margin(15, 30, 15, 15)
      ) +
    coord_capped_cart(bottom=capped_horizontal(c("both")),
                      left=capped_vertical(c("both"))) +
    guides(color = "none")
  print(p)
  
  
  # p <- ggplot(data = data,
  #        mapping = aes(x = reorder(snps, sort(as.numeric(positions))),
  #                      y = count, fill = snps)) +
  #   geom_bar(stat = "identity", position = "dodge") +
  #   geom_col() +
  #   geom_label(
  #     aes(label = perc),
  #     hjust = 1.0,
  #     size = 3, 
  #     fontface = "bold",
  #     fill = "#FFFFFF", 
  #     label.size = 0
  #   ) +
  #   # geom_text(aes(label = perc),
  #   #           hjust = 1.0,
  #   #           size = 3.5,
  #   #           fontface = "bold",
  #   #           color = "#000000") +
  #   # gghighlight(perc > 4, label_key = snps) +
  #   # geom_pointrange(aes(ymin = 0, ymax = count),
  #   #                 # Make the lines a little thicker and the dots a little bigger
  #   #                 fatten = 3, size = 1.0) +
  #   facet_grid(facets=. ~ host, labeller=labeller(drv=labels)) +
  #   scale_fill_discrete_diverging(palette = palette, rev = TRUE, alpha=1.0) +
  #   coord_flip() +
  #   xlab("Mutations") +
  #   ylab("No. of sequences") +
  #   ggtitle(paste(gene, ", vaccine: ", reference, sep = "")) +
  #   theme_classic(base_size = 30) +
  #   theme(
  #     text=element_text(family="Arial Narrow"),
  #     axis.line = element_line(colour = "#000000", linewidth = 0.4),
  #     axis.ticks = element_line(colour = "#000000", linewidth = 0.4),
  #     axis.title = element_text(colour = "#000000", face = "bold"),
  #     panel.grid.major = element_line(colour = "#B3B3B3", linewidth = 0.2),
  #     panel.grid.minor = element_blank(),
  #     panel.grid.major.y = element_blank(),
  #     panel.grid.minor.y = element_blank(),
  #     panel.spacing = unit(0.75, "cm", data = NULL),
  #     strip.background = element_blank(),
  #     strip.text.x = element_text(colour = "#000000", face = "bold", hjust = 0),
  #     plot.title = element_text(colour = "#000000", face = "bold"),
  #     axis.text = element_text(colour = "#000000", face = "bold"),
  #     #plot.tag = element_text(size = 32, colour = "#000000", face = "bold"),
  #     #axis.text.y = element_text(size = 20, colour = "#000000", face = "bold"),
  #     #axis.text.x = element_text(size = 20, colour = "#000000", face = "bold"),
  #     #axis.title = element_text(size = 22, colour = "#000000", face = "bold"),
  #     legend.position = "none",
  #     # legend.text = element_text(size = 18),
  #     # legend.title = element_text(size = 20),
  #     #legend.key = element_blank()
  #     ) +
  #   coord_capped_flip(bottom=capped_horizontal(c("both")),
  #                     left=capped_vertical(c("both"))) +
  #   guides(color = "none")
  # print(p)
  return(p)
}


get_common_substitutions <- function(file, snps_count, reference){
  #'
  #'
  #'
  #'
  #'
  
  df <- read.csv(file)
  df$host <- str_to_title(df$host)
  df1 <- df %>% 
    dplyr::group_by(host, snps, positions) %>% 
    dplyr::summarise(count=n()) 
  
  df2 <- df %>% dplyr::group_by(host, snps, positions) %>% 
    dplyr::summarise(count=n()) %>% 
    dplyr::arrange(desc(count))
  
  data_list <- list()
  df3 <- NULL
  for (i in 1:length(unique(df2$host))){
    df3$host <- unique(df2$host)[i]
    df3$total <- df2[df2$host == unique(df2$host)[i],]$count[[1]]
    data_list[[i]] <- df3
  }
  dat <- bind_rows(data_list)
  # get abundant snps
  data <- df1[df1$count > snps_count, ]
  data <- data[order(data$positions),]
  data <- dplyr::left_join(data, dat, by="host")
  data$percentage <- data$count/data$total*100
  data <- data |> dplyr::mutate_if(is.numeric, round, digits=1) 
  data <- data |> dplyr::mutate(perc = paste0(sprintf("%4.1f", count / total * 100), "%"))
  
  d <- data %>%
    dplyr::group_by(snps, positions) %>%
    dplyr::summarise(count=n()) %>%
    dplyr::filter(count >= 3) %>% 
    dplyr::arrange(desc(positions))
  d$vaccine <- reference
  return(d)
}

largeSubsMP12 <- get_common_substitutions(
  file = largeSingletonMP12File,
  snps_count = 10,
  reference = "MP-12"
)

largeSubsClone13 <- get_common_substitutions(
  file = largeSingletonClone13File,
  snps_count = 10,
  reference = "Clone-13"
)

largeSubsSmithburn <- get_common_substitutions(
  file = largeSingletonSmithburnFile,
  snps_count = 10,
  reference = "Smithburn")

dplyr::left_join(largeSubsSmithburn, 
                 largeSubsMP12, 
                 by="snps") |>
  dplyr::left_join(largeSubsClone13, by="snps") |>
  dplyr::filter(!is.na(positions.y))


mediumSubsMP12 <- get_common_substitutions(
  file = mediumSingletonMP12File,
  snps_count = 10,
  reference = "MP-12"
)

mediumSubsClone13 <- get_common_substitutions(
  file = mediumSingletonClone13File,
  snps_count = 10,
  reference = "Clone-13"
)

mediumSubsSmithburn <- get_common_substitutions(
  file = mediumSingletonSmithburnFile,
  snps_count = 10,
  reference = "Smithburn")

dplyr::left_join(mediumSubsSmithburn, 
                 mediumSubsMP12, 
                 by="snps") |>
  dplyr::left_join(mediumSubsClone13, by="snps") |>
  dplyr::filter(!is.na(positions.y))


nssSubsMP12 <- get_common_substitutions(
  file = nssSingletonMP12File,
  snps_count = 10,
  reference = "MP-12"
)

nssSubsClone13 <- get_common_substitutions(
  file = nssSingletonClone13File,
  snps_count = 10,
  reference = "Clone-13"
)

nssSubsSmithburn <- get_common_substitutions(
  file = nssSingletonSmithburnFile,
  snps_count = 10,
  reference = "Smithburn")

dplyr::left_join(nssSubsSmithburn, 
                 nssSubsMP12, 
                 by="snps") |>
  dplyr::left_join(nssSubsClone13, by="snps") |>
  dplyr::filter(!is.na(positions.y))

npSubsMP12 <- get_common_substitutions(
  file = npSingletonMP12File,
  snps_count = 3,
  reference = "MP-12"
)

npSubsClone13 <- get_common_substitutions(
  file = npSingletonClone13File,
  snps_count = 1,
  reference = "Clone-13"
)

npSubsSmithburn <- get_common_substitutions(
  file = npSingletonSmithburnFile,
  snps_count = 3,
  reference = "Smithburn")

dplyr::left_join(npSubsSmithburn, 
                 npSubsMP12, 
                 by="snps") |>
  dplyr::left_join(npSubsClone13, by="snps") |>
  dplyr::filter(!is.na(positions.y))

plotMutationsLargeMP12 <- plot_mutations(file = largeSingletonMP12File,
                                         snps_count = 10,
                                         reference = "MP-12",
                                         gene = "RdRp",
                                         palette = "Tofino"
                                         )
plotMutationsLargeClone13 <- plot_mutations(file = largeSingletonClone13File,
                                         snps_count = 10,
                                         reference = "Clone-13",
                                         gene = "RdRp",
                                         palette = "Berlin"
                                         )
plotMutationsLargeSmithburn <- plot_mutations(file = largeSingletonSmithburnFile,
                                            snps_count = 10,
                                            reference = "Smithburn",
                                            gene = "RdRp",
                                            palette = "Vik"
                                            )

plotMutationsMediumMP12 <- plot_mutations(file = mediumSingletonMP12File,
                                         snps_count = 10,
                                         reference = "MP-12",
                                         gene = "NSm/Gn/Gc",
                                         palette = "Tofino"
)
plotMutationsMediumClone13 <- plot_mutations(file = mediumSingletonClone13File,
                                            snps_count = 10,
                                            reference = "Clone-13",
                                            gene = "NSm/Gn/Gc",
                                            palette = "Berlin"
)
plotMutationsMediumSmithburn <- plot_mutations(file = mediumSingletonSmithburnFile,
                                              snps_count = 10,
                                              reference = "Smithburn",
                                              gene = "NSm/Gn/Gc",
                                              palette = "Vik"
)

plotMutationsNssMP12 <- plot_mutations(file = nssSingletonMP12File,
                                       snps_count = 10,
                                       reference = "MP-12",
                                       gene = "NSs",
                                       palette = "Tofino"
)
plotMutationsNssClone13 <- plot_mutations(file = nssSingletonClone13File,
                                          snps_count = 30,
                                          reference = "Clone-13",
                                          gene = "NSs",
                                          palette = "Berlin"
)
plotMutationsNssSmithburn <- plot_mutations(file = nssSingletonSmithburnFile,
                                            snps_count = 10,
                                            reference = "Smithburn",
                                            gene = "NSs",
                                            palette = "Vik"
)

plotMutationsNpMP12 <- plot_mutations(file = npSingletonMP12File,
                                      snps_count = 3,
                                      reference = "MP-12",
                                      gene = "NP",
                                      palette = "Tofino"
)
plotMutationsNpClone13 <- plot_mutations(file = npSingletonClone13File,
                                         snps_count = 1,
                                         reference = "Clone-13",
                                         gene = "NP",
                                         palette = "Berlin"
)
plotMutationsNpSmithburn <- plot_mutations(file = npSingletonSmithburnFile,
                                           snps_count = 3,
                                           reference = "Smithburn",
                                           gene = "NP",
                                           palette = "Vik"
)

mutations.barplot.MP12 <- ((plotMutationsLargeMP12 / 
                              plotMutationsMediumMP12) | 
                             (plotMutationsNssMP12/ 
                                plotMutationsNpMP12)) +
  plot_layout(guides = 'collect', 
              widths = c(5,5),
              heights = c(7,5)) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = 32, face = "bold", colour = "black"),
        strip.text.x = element_text(size = 28, colour = "#000000", face="bold"),
  )

ggsave("mutation-barplots-MP12.pdf", 
       mutations.barplot.MP12, 
       width = 40, height = 37, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)

ggsave("mutation-barplots-MP12.png", 
       mutations.barplot.MP12, 
       width = 40, height = 37, units = "in", 
       limitsize = FALSE,
       dpi = 600, bg="white", device = "png")

mutations.barplot.Clone13 <- ((plotMutationsLargeClone13 / 
                             plotMutationsMediumClone13) | 
                             (plotMutationsNpClone13/ 
                             plot_spacer())) +
  plot_layout(guides = 'collect', 
              widths = c(5, 5),
              heights = c(7,5,3,1)) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = 32, face = "bold", colour = "black"),
        axis.text = element_text(size = 28, face = "bold", colour = "black"),
        strip.text.x = element_text(size = 28, colour = "#000000", face="bold"),
  )

ggsave("mutation-barplots-Clone13.pdf", 
       mutations.barplot.Clone13, 
       width = 40, height = 37, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)

ggsave("mutation-barplots-Clone13.png", 
       mutations.barplot.Clone13, 
       width = 40, height = 37, units = "in", 
       limitsize = FALSE,
       dpi = 600, bg="white", device = "png")

mutations.barplot.Smithburn <- ((plotMutationsLargeSmithburn / 
                                   plotMutationsMediumSmithburn) | 
                                  (plotMutationsNssSmithburn/ 
                                     plotMutationsNpSmithburn)) +
  plot_layout(guides = 'collect', 
              widths = c(5, 5),
              heights = c(7,5,3,1)) +
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = 32, face = "bold", colour = "black"),
        axis.text = element_text(size = 28, face = "bold", colour = "black"),
        strip.text.x = element_text(size = 28, colour = "#000000", face="bold"),
  )

ggsave("mutation-barplots-Smithburn.pdf", 
       mutations.barplot.Smithburn, 
       width = 40, height = 37, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)

ggsave("mutation-barplots-Smithburn.png", 
       mutations.barplot.Smithburn, 
       width = 40, height = 37, units = "in", 
       limitsize = FALSE,
       dpi = 600, bg="white", device = "png")

mutation.barplots <- (
  (plotMutationsLargeMP12 | plotMutationsLargeClone13 | plotMutationsLargeSmithburn) / 
    (plotMutationsMediumMP12 | plotMutationsMediumClone13 | plotMutationsMediumSmithburn) / 
    (plotMutationsNssMP12 | plotMutationsNssClone13 | plotMutationsNssSmithburn) / 
    (plotMutationsNpMP12 | plotMutationsNpClone13 | plotMutationsNpSmithburn)
) +
  plot_layout(guides = 'collect', widths = c(10, 10, 24)) +
  plot_annotation(tag_levels = "A")
mutation.barplots

ggsave("mutation-barplots.pdf", mutation.barplots, 
       width = 45, height = 45, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)

ggsave("mutation-barplots.png", mutation.barplots, 
       width = 45, height = 60, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")




#################### glycosylation plots ######################

prepare_netglyc_data <- function(segment, indir, gene){
  
  fns <- file.path(indir, list.files(path = indir, 
                                     pattern = "\\.parsed.netglyc.csv$", 
                                     recursive = TRUE))
  data_list <- list()
  prefixes <- list()
  for (i in 1:length(fns)) {
    fn = fns[i]
    sample_name <- str_to_title(strsplit(basename(fn), ".", fixed=T)[[1]][3])
    df <- read.csv(fn, sep=',', stringsAsFactors = F, header = T)
    # df$Host <- sample_name
    data_list[[i]] <- df
  }
  data <- bind_rows(data_list)
  data$Host <- str_to_title(data$Host)
  cols <- c("Position", "Potential", "Agreement", "Result")
  data[cols] <- lapply(data[cols], factor)
  return(data)
}


netglyc.m <- prepare_netglyc_data(segment = "M",
                                  indir = "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/riftM/netglyc",
                                  gene = "NSm/Gn/Gc")

netglyc.np <- prepare_netglyc_data(segment = "NP",
                                   indir = "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/riftS-NP/netglyc",
                                   gene = "NP")

netglyc.l <- prepare_netglyc_data(segment = "L",
                                  indir = "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/riftL/netglyc",
                                  gene = "RdRp")

host.cols <- data.frame(Host=sort(unique((c(unique(netglyc.l$Host),
                                            unique(netglyc.m$Host),
                                            unique(netglyc.np$Host))))), 
                        Color=sequential_hcl(7, "Batlow", alpha = 1))

pot.cols <- data.frame(Potential=sort(unique((c(unique(netglyc.l$Potential),
                                                unique(netglyc.m$Potential),
                                                unique(netglyc.np$Potential))))), 
                       Color=divergingx_hcl(24, "Geyser"))


plot_netglyc <- function(dataset, min_genomes, segment, gene){
  #'
  #'
  #'
  #'
  #'
  summary <- dataset %>% 
    dplyr::group_by(Host,Accession) %>% 
    dplyr::summarise(Count=n()) %>% 
    dplyr::group_by(Host) %>% 
    dplyr::summarise(Genomes=n())
  
  dataset.potential <- dataset[dataset$Result %in% c('+', '++'),] %>%
    dplyr::group_by(Position, Host, Potential) %>%
    dplyr::summarise(Score=mean(Jury), Genomes=n())

  
  print(sort(unique(dataset.potential$Potential)))
  # Get only events with > 1 genome
  dt <- dataset.potential[dataset.potential$Genomes > min_genomes,]
  
  host.colors <- host.cols[host.cols$Host %in% sort(unique(dt$Host)),]$Color
  potential.colors <- pot.cols[pot.cols$Potential %in% sort(unique(dt$Potential)),]$Color
  print(sort(unique(dt$Potential)))
  
  p1 <- ggplot(dt, 
               aes(x=Position, y=Genomes, size=Genomes, fill=Host)) +
    geom_point(alpha=1.0, shape=21) +
    scale_size(range = c(1, 20)) +
    scale_fill_manual(name = "Host",
                      values = host.colors,
                      breaks = sort(unique(dt$Host)),
                      na.value = NA,
                      na.translate = F,
                      guide = guide_legend(override.aes = list(linetype = 0,
                                                               shape = 16,
                                                               color = host.colors))) +
    coord_flip() +
    xlab("Amino acid position") +
    ylab("Number of genomes") +
    ggtitle("", subtitle=paste0(segment, ": ", gene,  sep = " ")) +
    theme_classic(base_size = 30) +
    theme(
      text=element_text(family="Helvetica"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(colour = "#000000", linewidth = 0.4),
      axis.ticks.y = element_line(colour = "#000000", linewidth = 0.4),
      axis.ticks.x = element_line(colour = "#000000", linewidth = 0.4),
      plot.subtitle = element_text(colour = "#000000", face = "bold"),
      plot.tag = element_text(colour = "#000000", face = "bold"),
      axis.text = element_text(colour = "#000000", face = "bold"),
      axis.title = element_text(colour = "#000000", face = "bold"),
      legend.text = element_text(size = 20, face = "bold", colour = "black"),
      legend.title = element_text(size = 22, face = "bold", colour = "black"),
      legend.key = element_blank()) +
    coord_capped_flip(bottom=capped_horizontal(c("both")),
                      left=capped_vertical(c("both"))) +
    guides(fill = guide_legend(override.aes = list(size=15)))
  print(p1)
  
  p2 <- ggplot(dt,
               aes(x=Score, y=Position, size=Genomes,
                   colour=Host, fill=Potential)) +
    geom_point(alpha=1.0, size=15, show.legend = F) +
    scale_colour_manual(name = "Host",
                        values = host.colors,
                        breaks = sort(unique(dt$Host)),
                        na.value = NA,
                        na.translate = F,
                        guide = guide_legend(override.aes = list(linetype = 0,
                                                                 shape = 16,
                                                                 color = host.colors))) +
    scale_fill_manual(name = "Potential",
                      values = potential.colors,
                      breaks = sort(unique(dt$Potential)),
                      na.value = NA,
                      na.translate = F) +
    geom_label_repel(data = dt,
                     aes(label=Potential, fill=Potential),
                     size = 6,
                     colour = "#000000",
                     fontface = "bold",
                     nudge_x = 0.25,
                     nudge_y = 0.25,
                     segment.curvature = -0.1,
                     segment.ncp = 1.5,
                     show.legend = FALSE,
                     max.overlaps = 20,
                     angle = 0) +
    ggtitle("", subtitle = paste0(segment, ": ", gene)) +
    xlab("N-glycosylation Score") +
    ylab("Amino acid position") +
    geom_vline(xintercept=0.5, linetype="dashed", color = "#545454") +
    theme_classic(base_size = 30) +
    theme(
      text=element_text(family="Helvetica"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(colour = "#000000", linewidth = 0.4),
      axis.ticks.y = element_line(colour = "#000000", linewidth = 0.4),
      axis.ticks.x = element_line(colour = "#000000", linewidth = 0.4),
      plot.subtitle = element_text(colour = "#000000", face = "bold"),
      plot.tag = element_text(colour = "#000000", face = "bold"),
      axis.text = element_text(colour = "#000000", face = "bold"),
      axis.title = element_text(colour = "#000000", face = "bold"),
      legend.text = element_text(size = 20, face = "bold", colour = "black"),
      legend.title = element_text(size = 22, face = "bold", colour = "black"),
      legend.key = element_blank()) +
    coord_capped_cart(bottom=capped_horizontal(c("both")),
                      left=capped_vertical(c("both"))) +
    guides(colour = guide_legend(nrow = 12, override.aes = list(size = 15)),
           fill = guide_legend(override.aes = aes(label="")))
  print(p2)
  
  p3 <- ggplot(dt, aes(fill=Host, y=Genomes, x=Host)) +
    geom_bar(position="dodge", stat="identity", alpha=1.0) +
    scale_fill_manual(name = "Host",
                      values = host.colors,
                      breaks = sort(unique(dt$Host)),
                      na.value = NA,
                      na.translate = F,
                      guide = guide_legend(override.aes = list(linetype = 0,
                                                               shape = 16,
                                                               color = host.colors))) +
    geom_label(data=dt,
               aes(label = Genomes), 
               vjust = 1, 
               nudge_y = -.5,
               size = 7, 
               fontface = "bold", 
               family = "Arial Narrow",
               ## turn into white box without outline
               fill = "white", label.size = 0
    ) +
    ggtitle("", subtitle=paste0(segment, ": ", gene,  sep = " ")) +
    facet_wrap(~Potential, scales = "free") +
    theme(legend.position="none") +
    xlab("") +
    theme_classic(base_size = 30) +
    theme(
      text=element_text(family="Helvetica"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_blank(),
      strip.text.x = element_text(hjust = 0),
      axis.line = element_line(colour = "#000000", linewidth = 0.4),
      axis.ticks.y = element_line(colour = "#000000", linewidth = 0.4),
      axis.ticks.x = element_line(colour = "#000000", linewidth = 0.4),
      plot.subtitle = element_text(colour = "#000000", face = "bold"),
      plot.tag = element_text(colour = "#000000", face = "bold"),
      axis.text = element_text(colour = "#000000", face = "bold"),
      axis.text.x = element_text(angle = 90, face = "bold", colour = "black"),
      axis.title = element_text(colour = "#000000", face = "bold"),
      legend.text = element_text(size = 20, face = "bold", colour = "black"),
      legend.title = element_text(size = 22, face = "bold", colour = "black"),
      legend.key = element_blank()) +
    guides(fill = guide_legend(nrow = 12, override.aes = list(size = 15)))
  print(p3)
  
  
  datalist <- list()
  for (i in sort(unique(dt$Potential))){
    df <- dt[dt$Potential == i,]
    df$fraction <- df$Genomes / sum(summary$Genomes)
    df$ymax <- cumsum(df$fraction)
    df$ymin <- c(0, head(df$ymax, n=-1))
    df$labelPosition <- (df$ymax + df$ymin) / 2
    df$label <- paste0(df$Host, ": ", df$Genomes)
    datalist[[i]] <- df
  }
  
  data <- bind_rows(datalist)
  
  p4 <- ggplot(data, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=Host)) +
    geom_rect() +
    facet_wrap(~Potential, scales = "free") +
    geom_text(x=2, 
              aes(y=labelPosition, label=label, color=Host),
              size=4.5,
              fontface="bold",
              angle=0) + # x here controls label position (inner / outer)
    scale_colour_manual(name = "Host",
                        values = host.colors,
                        breaks = sort(unique(dt$Host)),
                        na.value = NA,
                        na.translate = F,
                        guide = guide_legend(override.aes = list(linetype = 0,
                                                                 shape = 16,
                                                                 color = host.colors))) +
    ggtitle("", subtitle=paste0(segment, ": ", gene,  sep = " ")) +
    scale_fill_manual(name = "Host",
                      values = host.colors,
                      breaks = sort(unique(dt$Host)),
                      na.value = NA,
                      na.translate = F,
                      guide = guide_legend(override.aes = list(linetype = 0,
                                                               shape = 16,
                                                               fill = host.colors))) +
    coord_polar(theta="y") +
    xlim(c(-1, 4)) +
    theme_void(base_size = 30) +
    theme(text=element_text(family="Arial Narrow"),
          legend.position = "none")
  print(p4)
  res <- list(p1,p2,p3,p4)
  return(res)
}



res.m <- plot_netglyc(dataset = netglyc.m, 
                      min_genomes = 1,
                      segment = "Medium", 
                      gene = "NSm/Gn/Gc")

res.np <- plot_netglyc(dataset = netglyc.np, 
                       min_genomes = 1,
                       segment = "Small", 
                       gene = "NP")

res.l <- plot_netglyc(dataset = netglyc.l,
                      min_genomes = 1,
                      segment = "Large", 
                      gene = "RdRp")



glyc.plots <- (((res.l[[1]] + theme(legend.position = "none"))  + 
                  (res.m[[1]] + theme(axis.title.y = element_blank()
                                      )) + 
                  (res.np[[1]] + theme(legend.position = "none",
                                                                                                                                              axis.title.y = element_blank()))) + 
                 plot_layout(guides = 'collect'))  /
  (((res.l[[2]]) + 
      (res.m[[2]] + theme(axis.title.y = element_blank())) + 
      (res.np[[2]] + theme(axis.title.y = element_blank()))) + 
     plot_layout(guides = 'collect')) +
  plot_layout(guides = 'collect', 
              widths = c(5, 5, 5),
              heights = c(5, 5)) +
  plot_annotation(tag_levels = "A") +
  theme(plot.tag = element_text(size = 32, face = "bold", colour = "black")
  )


ggsave("N-glycosylation-plots.pdf", glyc.plots, 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)


ggsave("N-glycosylation-plots.png", glyc.plots, 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")


ggsave("N-glycosylation-plot-Large.pdf", res.l[[3]], 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)


ggsave("N-glycosylation-plot-Large.png", res.l[[3]], 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")


ggsave("N-glycosylation-plot-Medium.pdf", res.m[[3]], 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)


ggsave("N-glycosylation-plot-Medium.png", res.m[[3]], 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")

ggsave("N-glycosylation-plot-NP.pdf", res.np[[3]], 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = cairo_pdf)


ggsave("N-glycosylation-plot-NP.png", res.np[[3]], 
       width = 30, height = 25, units = "in", 
       limitsize = FALSE,
       dpi = 300, bg="white", device = "png")


########################### plot sequence contexts ##########################

prepare_seq_vector <- function(segment, indir, mutation, prefix, reference, gene, hosts){
  
  pattern = paste0("\\.", mutation, ".", prefix, ".sequence-context.txt$")
  print(pattern)
  
  fns <- file.path(indir, list.files(path = indir, 
                                     pattern = pattern, 
                                     recursive = TRUE))
  data_list <- list()
  prefixes <- list()
  for (i in 1:length(fns)) {
    fn = fns[i]
    sample_name <- str_to_title(strsplit(basename(fn), ".", fixed=T)[[1]][1])
    df <- read.csv(fn, sep=',', header = F)
    ls <- list(as.character(df[,2:dim(df)[2]]))
    names(ls) <- sample_name
    prefixes[i] <- sample_name
    data_list[i] <- ls
  }
  names(data_list) <- prefixes

  d_list <- data_list[hosts]
  
  print(d_list)

  p <- ggplot() +
    geom_logo(d_list, method = "prob", seq_type = "dna") +
    xlab("nt") +
    facet_wrap(~seq_group, scales='free_x', ncol = 3) +
    scale_x_discrete(limits=c("-2", "-1", "0", "+1", "+2")) +
    ggtitle("", subtitle = paste0(segment, ": ", gene)) +
    theme_logo() +
    theme_classic(base_size = 30) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          text = element_text(family = "Helvetica"),
          plot.subtitle = element_text(size = 30, colour = "#000000", face = "bold"),
          strip.background = element_blank(),
          strip.text.x = element_text(hjust = 0, colour = "#000000", face = "bold.italic"),
          plot.tag = element_text(size = 32, colour = "#000000", face = "bold"),
          axis.line = element_line(colour = "#000000", linewidth = 0.4),
          axis.ticks = element_line(colour = "#000000", linewidth = 0.4),
          axis.text = element_text(size = 24, colour = "#000000", face = "bold", color = "black"),
          axis.title.x = element_blank(),
          axis.text.y = element_text(size = 24, hjust = 1, vjust = 0.5, face = "bold", colour = "black"),
          axis.title.y = element_text(size = 25, "#000000", face = "bold", colour = "black"),
          legend.key = element_blank()) +
    coord_capped_cart(bottom=capped_horizontal(c("both")),
                      left=capped_vertical(c("both")))
  return(p)
}


# large
ga.large.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftL/sequence-contexts"
ga.large.plus.plot <- prepare_seq_vector(segment = "Large",
                                         indir = ga.large.plus,
                                         mutation = "GA",
                                         prefix = "riftL",
                                         reference = "ZH-548",
                                         gene = "RdRp, G-A",
                                         hosts = c("Human", "Cow", "Sheep"))
ga.large.plus.plot <- ga.large.plus.plot + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank())
ga.large.plus.plot

ct.large.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftL/sequence-contexts"
ct.large.minus.plot <- prepare_seq_vector(segment = "Large",
                                           indir = ct.large.minus,
                                           mutation = "CT",
                                           prefix = "riftL",
                                           reference = "ZH-548",
                                           gene = "RdRp, C-T",
                                          hosts = c("Human", "Cow", "Sheep"))
ct.large.minus.plot <- ct.large.minus.plot + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank(),
  axis.title.y = element_blank(),
  axis.text.y = element_blank())
ct.large.minus.plot

# medium
ga.medium.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftM/sequence-contexts"
ga.medium.plus.plot <- prepare_seq_vector(segment = "Medium",
                                          indir = ga.medium.plus,
                                          mutation = "GA",
                                          prefix = "riftM",
                                          reference = "ZH-548",
                                          gene = "NSm/Gn/Gc, G-A",
                                          hosts = c("Human", "Cow", "Sheep"))
ga.medium.plus.plot <- ga.medium.plus.plot + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank())
ga.medium.plus.plot

ct.medium.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftM/sequence-contexts"
ct.medium.minus.plot <- prepare_seq_vector(segment = "Medium",
                                           indir = ct.medium.minus,
                                           mutation = "CT",
                                           prefix = "riftM",
                                           reference = "ZH-548",
                                           gene = "NSm/Gn/Gc, C-T",
                                           hosts = c("Human", "Cow", "Sheep"))
ct.medium.minus.plot <- ct.medium.minus.plot + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank(),
  axis.title.y = element_blank(),
  axis.text.y = element_blank())
ct.medium.minus.plot

# NSs
ga.nss.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NSS/sequence-contexts"
ga.nss.plus.plot <- prepare_seq_vector(segment = "Small",
                                       indir = ga.nss.plus,
                                       mutation = "GA",
                                       prefix = "riftS-NSS",
                                       reference = "ZH-548",
                                       gene = "NSs, G-A",
                                       hosts = c("Human", "Cow", "Sheep"))
ga.nss.plus.plot <- ga.nss.plus.plot + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank())
ga.nss.plus.plot

ct.nss.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NSS/sequence-contexts"
ct.nss.minus.plot <- prepare_seq_vector(segment = "Small",
                                        indir = ct.nss.minus,
                                        mutation = "CT",
                                        prefix = "riftS-NSS",
                                        reference = "ZH-548",
                                        gene = "NSs, C-T",
                                        hosts = c("Human", "Cow", "Sheep"))
ct.nss.minus.plot <- ct.nss.minus.plot + theme(
  axis.title.x = element_blank(),
  axis.text.x = element_blank(),
  axis.title.y = element_blank(),
  axis.text.y = element_blank())
ct.nss.minus.plot


# np
ga.np.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NP/sequence-contexts"
ga.np.plus.plot <- prepare_seq_vector(segment = "Small",
                                      indir = ga.np.plus,
                                      mutation = "GA",
                                      prefix = "riftS-NP",
                                      reference = "ZH-548",
                                      gene = "NP, G-A",
                                      hosts = c("Human", "Cow", "Sheep"))
# ga.np.plus.plot <- ga.np.plus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ga.np.plus.plot

ct.np.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NP/sequence-contexts"
ct.np.minus.plot <- prepare_seq_vector(segment = "Small",
                                       indir = ct.np.minus,
                                       mutation = "CT",
                                       prefix = "riftS-NP",
                                       reference = "ZH-548",
                                       gene = "NP, C-T",
                                       hosts = c("Human", "Cow", "Sheep"))
ct.np.minus.plot <- ct.np.minus.plot + theme(
  axis.title.y = element_blank(),
  axis.text.y = element_blank())
ct.np.minus.plot

seq.contexts.plots <- 
  ((ga.large.plus.plot | ct.large.minus.plot) / 
  (ga.medium.plus.plot | ct.medium.minus.plot) / 
  (ga.nss.plus.plot | ct.nss.minus.plot) / 
  (ga.np.plus.plot | ct.np.minus.plot)) +
  plot_annotation(tag_levels = "A") +
  plot_layout(guides = 'collect',
              widths = c(2,2),
              heights = c(3,3,3,3))

ggsave("sequence-context-plots.pdf", 
       seq.contexts.plots, 
       width = 24, 
       height = 25, 
       units = "in", 
       limitsize = FALSE,
       dpi = 300, 
       bg="white", 
       device = cairo_pdf)

ggsave("sequence-context-plots.png", 
       seq.contexts.plots, 
       width = 24, 
       height = 25, 
       units = "in", 
       limitsize = FALSE,
       dpi = 300, 
       bg="white", 
       device = "png")

# large.seq.contexts.plots <- 
#   ((ga.large.plus.plot) | (ct.large.minus.plot + 
#                             theme(axis.title.y = element_blank(),
#                                   axis.text.y = element_blank())))+
#   plot_layout(widths = c(5,5), heights = c(6,6)) +
#   plot_annotation(tag_levels = "A")
# 
# 
# ggsave("large-sequence-context-plots.pdf", 
#        large.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = cairo_pdf)
# 
# ggsave("large-sequence-context-plots.png", 
#        large.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = "png")
# 
# 
# medium.seq.contexts.plots <- 
#   ((ga.medium.plus.plot) | (ct.medium.minus.plot + 
#                               theme(axis.title.y = element_blank(),
#                                     axis.text.y = element_blank())))+
#   plot_layout(widths = c(5,5), heights = c(6,6)) +
#   plot_annotation(tag_levels = "A")
# 
# 
# ggsave("medium-sequence-context-plots.pdf", 
#        medium.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = cairo_pdf)
# 
# ggsave("medium-sequence-context-plots.png", 
#        medium.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = "png")
# 
# 
# nss.seq.contexts.plots <- 
#   ((ga.nss.plus.plot) | (ct.nss.minus.plot + 
#                            theme(axis.title.y = element_blank(),
#                                  axis.text.y = element_blank())))+
#   plot_layout(widths = c(5,5), heights = c(6,6)) +
#   plot_annotation(tag_levels = "A")
# 
# 
# ggsave("nss-sequence-context-plots.pdf", 
#        nss.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = cairo_pdf)
# 
# ggsave("nss-sequence-context-plots.png", 
#        nss.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = "png")
# 
# np.seq.contexts.plots <- 
#   ((ga.np.plus.plot) | (ct.np.minus.plot + 
#                           theme(axis.title.y = element_blank(),
#                                 axis.text.y = element_blank())))+
#   plot_layout(widths = c(5,5), heights = c(6,6)) +
#   plot_annotation(tag_levels = "A")
# 
# 
# ggsave("np-sequence-context-plots.pdf", 
#        np.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = cairo_pdf)
# 
# ggsave("np-sequence-context-plots.png", 
#        np.seq.contexts.plots, 
#        width = 28, 
#        height = 30, 
#        units = "in", 
#        limitsize = FALSE,
#        dpi = 300, 
#        bg="white", 
#        device = "png")

prepare_seq_vector_apobecs <- function(segment, indir, mutation, prefix, reference, gene, hosts){
  
  pattern = paste0("\\.", mutation, ".", prefix, ".context.txt$")
  print(pattern)
  
  fns <- file.path(indir, list.files(path = indir, 
                                     pattern = pattern, 
                                     recursive = TRUE))
  data_list <- list()
  prefixes <- list()
  for (i in 1:length(fns)) {
    fn = fns[i]
    sample_name <- strsplit(basename(fn), ".", fixed=T)[[1]][1]
    print(sample_name)
    df <- read.csv(fn, sep=',', header = F)
    ls <- list(as.character(df[,2:dim(df)[2]]))
    print(ls)
    names(ls) <- sample_name
    prefixes[i] <- sample_name
    data_list[i] <- ls
  }
  names(data_list) <- prefixes

  d_list <- data_list[hosts]
  
  seq_group.labs <- c("A3A/A3B", "A3C/A3F")
  names(seq_group.labs) <- c("A3A_A3B", "A3C_A3F")

  p <- ggplot() +
    geom_logo(d_list, method = "prob", seq_type = "dna") +
    xlab("nt") +
    facet_wrap(~seq_group, ncol=3, scales='free_x',
               labeller = labeller(seq_group = seq_group.labs)) +
    scale_x_discrete(limits=c("-2", "-1", "0", "+1", "+2")) +
    ggtitle("", subtitle = paste0(segment, ": ", gene)) +
    theme_logo() +
    theme_classic(base_size = 24) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          text = element_text(family = "Helvetica"),
          plot.subtitle = element_text(size = 24, colour = "#000000", face = "bold"),
          strip.background = element_blank(),
          strip.text.x = element_text(hjust = 0, colour = "#000000", face = "bold"),
          plot.tag = element_text(size = 32, colour = "#000000", face = "bold"),
          axis.line = element_line(colour = "#000000", linewidth = 0.4),
          axis.ticks = element_line(colour = "#000000", linewidth = 0.4),
          axis.text = element_text(size = 16, colour = "#000000", face = "bold"),
          axis.title.x = element_blank(),
          axis.text.y = element_text(size = 16, hjust = 1, vjust = 0.5),
          axis.title.y = element_text(size = 18, "#000000", face = "bold"),
          legend.key = element_blank()) +
    coord_capped_cart(bottom=capped_horizontal(c("both")),
                      left=capped_vertical(c("both")))
  return(p)
}


# large
ga.large.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftL/sequence-contexts"
ga.large.plus.plot <- prepare_seq_vector_apobecs(segment = "Large",
                                                 indir = ga.large.plus,
                                                 mutation = "GA",
                                                 prefix = "riftL",
                                                 reference = "ZH-548",
                                                 gene = "RdRp, G-A",
                                                 hosts = c("A3A_A3B", "A3C_A3F"))
# ga.large.plus.plot <- ga.large.plus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ga.large.plus.plot

ct.large.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftL/sequence-contexts"
ct.large.minus.plot <- prepare_seq_vector_apobecs(segment = "Large",
                                                  indir = ct.large.minus,
                                                  mutation = "CT",
                                                  prefix = "riftL",
                                                  reference = "ZH-548",
                                                  gene = "RdRp, C-T",
                                                  hosts = c("A3A_A3B", "A3C_A3F"))
# ct.large.minus.plot <- ct.large.minus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ct.large.minus.plot <- ct.large.minus.plot + 
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank())

# medium
ga.medium.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftM/sequence-contexts"
ga.medium.plus.plot <- prepare_seq_vector_apobecs(segment = "Medium",
                                                  indir = ga.medium.plus,
                                                  mutation = "GA",
                                                  prefix = "riftM",
                                                  reference = "ZH-548",
                                                  gene = "NSm/Gn/Gc, G-A",
                                                  hosts = c("A3G", "A3A_A3B", "A3C_A3F"))
# ga.medium.plus.plot <- ga.medium.plus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ga.medium.plus.plot

ct.medium.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftM/sequence-contexts"
ct.medium.minus.plot <- prepare_seq_vector_apobecs(segment = "Medium",
                                                   indir = ct.medium.minus,
                                                   mutation = "CT",
                                                   prefix = "riftM",
                                                   reference = "ZH-548",
                                                   gene = "NSm/Gn/Gc, C-T",
                                                   hosts = c("A3G", "A3A_A3B", "A3C_A3F"))
# ct.medium.minus.plot <- ct.medium.minus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ct.medium.minus.plot <- ct.medium.minus.plot + 
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank())

# NSs
ga.nss.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NSS/sequence-contexts"
ga.nss.plus.plot <- prepare_seq_vector_apobecs(segment = "Small",
                                               indir = ga.nss.plus,
                                               mutation = "GA",
                                               prefix = "riftS-NSS",
                                               reference = "ZH-548",
                                               gene = "NSs, G-A",
                                               hosts = c("A3A_A3B"))
# ga.nss.plus.plot <- ga.nss.plus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ga.nss.plus.plot

ct.nss.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NSS/sequence-contexts"
ct.nss.minus.plot <- prepare_seq_vector_apobecs(segment = "Small",
                                                indir = ct.nss.minus,
                                                mutation = "CT",
                                                prefix = "riftS-NSS",
                                                reference = "ZH-548",
                                                gene = "NSs, C-T",
                                                hosts = c("A3A_A3B"))
# ct.nss.minus.plot <- ct.nss.minus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ct.nss.minus.plot <- ct.nss.minus.plot + 
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank())

# np
ga.np.plus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NP/sequence-contexts"
ga.np.plus.plot <- prepare_seq_vector_apobecs(segment = "Small",
                                      indir = ga.np.plus,
                                      mutation = "GA",
                                      prefix = "riftS-NP",
                                      reference = "ZH-548",
                                      gene = "NP, G-A",
                                      hosts = c("A3A_A3B", "A3C_A3F"))
# ga.np.plus.plot <- ga.np.plus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ga.np.plus.plot

ct.np.minus <- "/Users/jjuma/projects/pipelines/rvfvphylo/output-dir/mutational-profiling/riftS-NP/sequence-contexts"
ct.np.minus.plot <- prepare_seq_vector_apobecs(segment = "Small",
                                       indir = ct.np.minus,
                                       mutation = "CT",
                                       prefix = "riftS-NP",
                                       reference = "ZH-548",
                                       gene = "NP, C-T",
                                       hosts = c("A3A_A3B", "A3C_A3F"))
# ct.np.minus.plot <- ct.np.minus.plot + theme(
#   axis.title.x = element_blank(),
#   axis.text.x = element_blank())
ct.np.minus.plot <- ct.np.minus.plot + 
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank())




seq.contexts.plots.apobecs <- 
  ((ga.large.plus.plot | ct.large.minus.plot) / 
     (ga.medium.plus.plot | ct.medium.minus.plot) / 
     (ga.np.plus.plot | ct.np.minus.plot) /
     (ga.nss.plus.plot | ct.nss.minus.plot)) +
  plot_annotation(tag_levels = "A") +
  plot_layout(guides = 'collect',
              widths = c(2,2),
              heights = c(3,3,3,3)) +
  theme(plot.tag = element_text(size = 32,  face = "bold", colour = "black"))

# seq.contexts.plots.apobecs <- 
#   ((ga.large.plus.plot) / (ga.medium.plus.plot) |
#    (ct.large.minus.plot) / 
#      (ct.medium.minus.plot) /
#     (ga.nss.plus.plot) / (ga.np.plus.plot) |
#      (ct.nss.minus.plot) / 
#        (ct.np.minus.plot) +
#   plot_annotation(tag_levels = "A") +
#   plot_layout(widths = c(3, 3, 3), heights = c(5, 5))
seq.contexts.plots.apobecs

ggsave("sequence-context-plots-apobecs.pdf", 
       seq.contexts.plots.apobecs, 
       width = 27, 
       height = 20, 
       units = "in", 
       limitsize = FALSE,
       dpi = 300, 
       bg="white", 
       device = cairo_pdf)

ggsave("sequence-context-plots-apobecs.png", 
       seq.contexts.plots.apobecs, 
       width = 28, 
       height = 20, 
       units = "in", 
       limitsize = FALSE,
       dpi = 300, 
       bg="white", 
       device = "png")


